---
title: "vAMPirus DataCheck Report: `r commandArgs(trailingOnly=T)[1]`"
date: "Generated on: `r Sys.time()`"
output: html_document
params:
  interactive: TRUE
  projtag: !r commandArgs(trailingOnly=T)[1]
  skipReadProcessing: !r commandArgs(trailingOnly=T)[2]
  skipMerging: !r commandArgs(trailingOnly=T)[3]
  skipAdapterRemoval: !r commandArgs(trailingOnly=T)[4]
  asvMED: !r commandArgs(trailingOnly=T)[5]
  aminoMED: !r commandArgs(trailingOnly=T)[6]
---
<style>
.rectangle {
  height: 37px;
  width: 105%;
  background-color: #576675;
}
</style>

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE,
      message = FALSE,
      warning = FALSE,
      out.width="100%")
```

```{r pathways, echo=FALSE}
knitr::include_graphics("vamplogo.png")
```

```{r load_libraries, include=FALSE}

library(vegan)
#library(rstatix)
library(tidyverse)
library(scales)
library(cowplot)
library(dplyr)
library(plotly)
library(knitr)
library(kableExtra)
library(rmarkdown)
```

```{r colors, include=FALSE}
mycol=c('#088da5','#73cdc8','#ff6f61','#7cb8df','#88b04b','#00a199','#6B5B95','#92A8D1','#b0e0e6','#ff7f50','#088d9b','#E15D44','#e19336')
```
<br>
<br>
------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
NOTE: Plots are interactive and you can use the legend to specify samples/treatment of interest. You can also download an .svg version of each figure within this report.
------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
<br>
<div class="rectangle"><h2 style="color:white">&nbsp;&nbsp;Pre- and Post-Adapter Removal Read Stats</h2></div>
<br>
```{r readstats, echo=FALSE}
if (params$skipReadProcessing == "true" || params$skipMerging == "true" ) {
    writeLines("\n--------------------------------------------------------------\n")
    cat(readLines(list.files(pattern="filter_reads.txt")) , sep = '\n')
    writeLines("\n--------------------------------------------------------------\n")
} else {
    if (params$skipAdapterRemoval == "false") {
        reads_stats=read.csv("final_reads_stats.csv")
        paged_table(reads_stats,options = list(rows.print = 10))
    } else {
        # file fore here
        writeLines("\n--------------------------------------------------------------\n")
        cat(readLines(list.files(pattern="filter_reads.txt")) , sep = '\n')
        writeLines("\n--------------------------------------------------------------\n")
    }
}
```
<br>
<br>

### Total number of reads before and after adapter removal

```{r readstats_plot, echo=FALSE}
# Plot of reads before and after
if (params$skipReadProcessing == "true" || params$skipMerging == "true" ) {
    writeLines("\n--------------------------------------------------------------\n")
    cat(readLines(list.files(pattern="filter_reads.txt")) , sep = '\n')
    writeLines("\n--------------------------------------------------------------\n")
} else {
    if (params$skipAdapterRemoval == "false") {
        ptotal <- plot_ly(type="box",marker=list(colors=mycol))
        ptotal <- ptotal %>% add_boxplot(y=reads_stats$Total_before, name="Reads before filtering")
        ptotal <- ptotal %>% add_boxplot(y=reads_stats$Total_after, name="Reads after filtering")
        #ptotal <- ptotal %>% layout(title=list(text="Number of reads before and after filtering"))
        ptotal <- ptotal %>% layout(legend = list(x=10,y=.5))
        ptotal <- ptotal %>% config(toImageButtonOptions=list(format='svg',filename='reads_pre_post_adaptrem', height= 500, width= 800, scale= 1))
        ptotal
    } else {
        # file fore here
        writeLines("\n--------------------------------------------------------------\n")
        cat(readLines(list.files(pattern="filter_reads.txt")) , sep = '\n')
        writeLines("\n--------------------------------------------------------------\n")
    }
}
```
<br>

### Forward (R1) and reverse (R2) read length before and after adapter removal

```{r readstats_plot2, echo=FALSE}
# Plot of R1 and R2 before and after
if (params$skipReadProcessing == "true" || params$skipMerging == "true" ) {
    writeLines("\n--------------------------------------------------------------\n")
    cat(readLines(list.files(pattern="filter_reads.txt")) , sep = '\n')
    writeLines("\n--------------------------------------------------------------\n")
} else {
    if (params$skipAdapterRemoval == "false") {
        pr <- plot_ly(y=reads_stats$R1_before_length, type="box", name="R1 length before")
        pr <- pr %>% add_boxplot(y=reads_stats$R1_after_length, name="R1 length after")
        pr <- pr %>% add_boxplot(y=reads_stats$R2_before_length, name="R2 length before")
        pr <- pr %>% add_boxplot(y=reads_stats$R2_after_length, name="R2 length after")
        #pr <- pr %>% layout(title = "R1 and R2 Length")
        pr <- pr %>% layout(legend = list(x=10,y=.5))
        pr <- pr %>% config(toImageButtonOptions=list(format='svg',filename='readLen_pre_post_adaptrem', height= 500, width= 800, scale= 1))
        pr
    } else {
        # file fore here
        writeLines("\n--------------------------------------------------------------\n")
        cat(readLines(list.files(pattern="filter_reads.txt")) , sep = '\n')
        writeLines("\n--------------------------------------------------------------\n")
    }
}
```
<br>
<br>
<br>
<div class="rectangle"><h2 style="color:white">&nbsp;&nbsp;Post-Merging Read Stats</h2></div>
<br>
<br>

### Pre-filtering reads per sample

<br>

```{r preFilt_preClean_reads, echo=FALSE}
prer=read.csv("reads_per_sample_preFilt_preClean.csv", header=TRUE)
prerp <- plot_ly(prer, x = ~reads, y = ~sample, name = "Sample", type = 'scatter',
               mode = "markers", marker = list(color = "#088da5"),
               hovertemplate = paste('Sample: %{y}','<br>Total reads: %{x}','<extra></extra>'))
prerp <- prerp %>% layout(xaxis = list(title = "Total reads"),yaxis = list(title = "Sample"))
prerp <- prerp %>% config(toImageButtonOptions=list(format='svg',filename='ReadsPerSamp_prefilt', height= 500, width= 800, scale= 1))
prerp
```

<br>

### Post-filtering reads per sample

<br>

```{r postFilt_postClean_reads, echo=FALSE}
postr=read.csv("read_per_sample_postFilt_postClean.csv", header=TRUE)
postrp <- plot_ly(postr, x = ~reads, y = ~sample, name = "Sample", type = 'scatter',
               mode = "markers", marker = list(color = "#088da5"),
               hovertemplate = paste('Sample: %{y}','<br>Total reads: %{x}','<extra></extra>'))
postrp <- postrp %>% layout(xaxis = list(title = "Total reads"),yaxis = list(title = "Sample"))
postrp <- postrp %>% config(toImageButtonOptions=list(format='svg',filename='ReadsPerSamp_postfilt', height= 500, width= 800, scale= 1))
postrp
```

<br>

### Pre-filtering base frequency per position on reads

<br>

```{r preClean_baseFrequency, echo=FALSE}
prec=read.csv(paste(params$projtag,sep="","_all_merged_preFilt_preClean_baseFrequency_hist.csv"), header=TRUE)
#prec=read.csv("PVID_vAMPtest1_all_merged_preFilt_preClean_baseFrequency_hist.csv", header=TRUE)
precp <- plot_ly(x=c(1:nrow(prec)),y=as.numeric(prec[,2]), type="scatter", mode = "markers", name="A", opacity = .8, markers=list(color='#088da5'))
precp <- precp %>% add_trace(y=as.numeric(prec[,3]), name="C", opacity = .8, markers=list(color='#e19336'))
precp <- precp %>% add_trace(y=as.numeric(prec[,3]), name="G", opacity = .8, markers=list(color='#ff6f61'))
precp <- precp %>% add_trace(y=as.numeric(prec[,3]), name="T", opacity = .8, markers=list(color='#7cb8df'))
precp <- precp %>% add_trace(y=as.numeric(prec[,3]), name="N", opacity = .8, markers=list(color='#88b04b'))
precp <- precp %>% layout(xaxis=list(title="Base position"))
precp <- precp %>% layout(yaxis=list(title="Base Frequency", range = c(0,1)))
precp <- precp %>% layout(legend = list(x=10,y=.5), hovermode = "x unified")
precp <- precp %>% config(toImageButtonOptions=list(format='svg',filename='BaseFreq_prefilt', height= 500, width= 800, scale= 1))
precp
```

<br>

### Post-filtering base frequency per position on reads

<br>

```{r postClean_baseFrequency, echo=FALSE}
postc=read.csv(paste(params$projtag,sep="","_all_merged_postFilt_baseFrequency_hist.csv"), header=TRUE)
#postc=read.csv("PVID_vAMPtest1_all_merged_postFilt_baseFrequency_hist.csv", header=TRUE)
postcp <- plot_ly(x=c(1:nrow(postc)),y=as.numeric(postc[,2]),type="scatter", mode = "markers", name="A", opacity = .8, markers=list(color='#088da5'))
postcp <- postcp %>% add_trace(y=as.numeric(postc[,3]), name="C", opacity = .8, markers=list(color='#e19336'))
postcp <- postcp %>% add_trace(y=as.numeric(postc[,3]), name="G", opacity = .8, markers=list(color='#ff6f61'))
postcp <- postcp %>% add_trace(y=as.numeric(postc[,3]), name="T", opacity = .8, markers=list(color='#7cb8df'))
postcp <- postcp %>% add_trace(y=as.numeric(postc[,3]), name="N", opacity = .8, markers=list(color='#88b04b'))
postcp <- postcp %>% layout(xaxis=list(title="Base position"))
postcp <- postcp %>% layout(yaxis=list(title="Base Frequency", range = c(0,1)))
postcp <- postcp %>% layout(legend = list(x=10,y=.5), hovermode = "x unified")
postcp <- postcp %>% config(toImageButtonOptions=list(format='svg',filename='BaseFreq_postfilt', height= 500, width= 800, scale= 1))
postcp
```

<br>

### Pre-filtering mean quality score per position on reads

<br>

```{bash preClean_qualityScore_bash, echo=FALSE}
cat *_all_merged_preFilt_preClean_qualityScore_hist.csv | grep -v "#" >edit_all_merged_preFilt_preClean_qualityScore_hist.csv
```
```{r preClean_qualityScore, echo=FALSE}
#preq=read.csv(paste(params$projtag,sep="","_all_merged_preFilt_preClean_qualityScore_hist.csv", header=TRUE))
preq=read.csv("edit_all_merged_preFilt_preClean_qualityScore_hist.csv",header=F)
#preqp <- plot_ly(x=c(1:nrow(preq)),y=as.numeric(preq[,2]),type="scatter", mode = "lines", name="Reads", opacity = .8, line=list(color='#088da5'))
preqp <- plot_ly(preq,x=~V1, y=~V2,type="scatter", mode = "lines", name="Reads", opacity = .8, line=list(color='#088da5'))
preqp <- preqp %>% layout(xaxis=list(title="Base position"))
preqp <- preqp %>% layout(yaxis=list(title="Mean quality", range = c(0,41)))
preqp <- preqp %>% layout(legend = list(x=10,y=.5), hovermode = "x unified")
preqp <- preqp %>% config(toImageButtonOptions=list(format='svg',filename='Qualscore_prefilt', height= 500, width= 800, scale= 1))
preqp
```

<br>

### Post-filtering mean quality score per position on reads

<br>

```{bash postClean_qualityScore_bash, echo=FALSE}
cat *_all_merged_postFilt_qualityScore_hist.csv | grep -v "#" >edit_all_merged_postFilt_qualityScore_hist.csv
```
```{r postClean_qualityScore, echo=FALSE}
#podtq=read.csv(paste(params$projtag,sep="","_all_merged_postFilt_qualityScore_hist.csv", header=TRUE))
postq=read.csv("edit_all_merged_postFilt_qualityScore_hist.csv", header=FALSE)
#postqp <- plot_ly(x=c(1:nrow(postq)),y=as.numeric(postq[,2]),type="scatter", mode = "lines", name="Reads", opacity = .8, line=list(color='#088da5'))
postqp <- plot_ly(postq, x=~V1, y=~V2, type="scatter", mode = "lines", name="Reads", opacity = .8, line=list(color='#088da5'))
postqp <- postqp %>% layout(xaxis=list(title="Base position"))
postqp <- postqp %>% layout(yaxis=list(title="Mean quality", range = c(0,41)))
postqp <- postqp %>% layout(legend = list(x=10,y=.5), hovermode = "x unified")
postqp <- postqp %>% config(toImageButtonOptions=list(format='svg',filename='Qualscore_postfilt', height= 500, width= 800, scale= 1))
postqp

```

<br>

### Pre-filtering read GC-content

<br>

```{bash gc, echo=FALSE}
cat *all_merged_preFilt_preClean_gcContent_hist.csv | grep -v "#" >edit_all_merged_preFilt_preClean_gcContent_hist.csv
```
```{r gc_plot, echo=FALSE}
pregc=read.csv("edit_all_merged_preFilt_preClean_gcContent_hist.csv", header=F)
pregcp <- plot_ly(pregc, x=~V1, y=~V2, type="bar", marker=list(color='#088da5', line=list(color = 'black',
               width = .1)), hovertemplate = paste('GC content: %{x}','<br>Count: %{y}','<extra></extra>'))
pregcp <- pregcp %>% layout(yaxis = list(title = "Count"))
pregcp <- pregcp %>% layout(xaxis = list(title = "GC content"))
pregcp <- pregcp %>% config(toImageButtonOptions=list(format='svg',filename='GCcont_prefilt', height= 500, width= 800, scale= 1))
pregcp
```

<br>

### Post-filtering read GC-content

<br>

```{bash gc2, echo=FALSE}
cat *all_merged_postFilt_gcContent_hist.csv | grep -v "#" >edit_all_merged_postFilt_gcContent_hist.csv
```
```{r gc2_plot, echo=FALSE}
postgc=read.csv("edit_all_merged_postFilt_gcContent_hist.csv", header=F)
postgcp <- plot_ly(postgc, x=~V1, y=~V2, type="bar", marker=list(color='#088da5', line=list(color = 'black',
               width = .1)), hovertemplate = paste('GC content: %{x}','<br>Count: %{y}','<extra></extra>'))
postgcp <- postgcp %>% layout(yaxis = list(title = "Count"))
postgcp <- postgcp %>% layout(xaxis = list(title = "GC content"))
postgcp <- postgcp %>% config(toImageButtonOptions=list(format='svg',filename='GCcont_postfilt', height= 500, width= 800, scale= 1))
postgcp
```

<br>

### Number of reads per quality score (pre-filtering)

<br>

```{r pre_averageQuality_hist, echo=FALSE}
preaq=read.csv(paste(params$projtag,sep="","_all_merged_preFilt_preClean_averageQuality_hist.csv"), header=TRUE)
#preaq=read.csv("PVID_vAMPtest1_all_merged_preFilt_preClean_averageQuality_hist.csv", header=T)
preaqp <- plot_ly(preaq, x=preaq[,1], y=preaq[,2], type="bar", marker=list(color='#088da5', line=list(color = 'black',
               width = .1)), hovertemplate = paste('Average quality: %{x}','<br>Count: %{y}','<extra></extra>'))
preaqp <- preaqp %>% layout(yaxis = list(title = "Count"))
preaqp <- preaqp %>% layout(xaxis = list(title = "Average quality"))
preaqp <- preaqp %>% config(toImageButtonOptions=list(format='svg',filename='ReQualScore_prefilt', height= 500, width= 800, scale= 1))
preaqp
```

<br>

### Number of reads per quality score (post-filtering)

<br>


```{r post_averageQuality_hist, echo=FALSE}
#postaq=read.csv(params$postFilt_averageQuaulity, header=TRUE)
postaq=read.csv(paste(params$projtag,sep="","_all_merged_preFilt_preClean_averageQuality_hist.csv"), header=TRUE)
#postaq=read.csv("PVID_vAMPtest1_all_merged_postFilt_averageQuaulity_hist.csv", header=T)
postaqp <- plot_ly(postaq, x=postaq[,1], y=postaq[,2], type="bar", marker=list(color='#088da5', line=list(color = 'black',
               width = .1)), hovertemplate = paste('Average quality: %{x}','<br>Count: %{y}','<extra></extra>'))
postaqp <- postaqp %>% layout(yaxis = list(title = "Count"))
postaqp <- postaqp %>% layout(xaxis = list(title = "Average quality"))
postaqp <- postaqp %>% config(toImageButtonOptions=list(format='svg',filename='ReQualScore_postfilt', height= 500, width= 800, scale= 1))
postaqp
```


<br>

### Number of reads per length (pre-filtering)

<br>


```{r pre_length_hist, echo=FALSE}
prelh=read.csv(paste(params$projtag,sep="","_all_merged_preFilt_preClean_length_hist.csv"), header=TRUE)
#prelh=read.csv("PVID_vAMPtest1_all_merged_preFilt_preClean_length_hist.csv", header=T)
prelhp <- plot_ly(prelh, x=prelh[,1], y=prelh[,2], type="bar", marker=list(color='#088da5', line=list(color = 'black',
               width = .1)), hovertemplate = paste('Read Length: %{x}','<br>Count: %{y}','<extra></extra>'))
prelhp <- prelhp %>% layout(yaxis = list(title = "Count"))
prelhp <- prelhp %>% layout(xaxis = list(title = "Read Length"))
prelhp <- prelhp %>% config(toImageButtonOptions=list(format='svg',filename='ReadsperLen_prefilt', height= 500, width= 800, scale= 1))
prelhp
```

<br>

### Number of reads per length (post-filtering)

<br>

```{r post_length_hist, echo=FALSE}
postlh=read.csv(paste(params$projtag,sep="","_all_merged_postFilt_length_hist.csv"), header=TRUE)
#postlh=read.csv("PVID_vAMPtest1_all_merged_postFilt_length_hist.csv", header=T)
postlhp <- plot_ly(postlh, x=postlh[,1], y=postlh[,2], type="bar", marker=list(color='#088da5', line=list(color = 'black',
               width = .1)), hovertemplate = paste('Read Length: %{x}','<br>Count: %{y}','<extra></extra>'))
postlhp <- postlhp %>% layout(yaxis = list(title = "Count"))
postlhp <- postlhp %>% layout(xaxis = list(title = "Read Length"))
postlhp <- postlhp %>% config(toImageButtonOptions=list(format='svg',filename='ReadsperLen_postfilt', height= 500, width= 800, scale= 1))
postlhp
```
<br>
```{bash load_datasets_bash, include=FALSE}
mv *AminoType_PairwiseDistance.matrix ./amino_matrix.txt
mv *_ASV_PairwiseDistance.matrix ./asv_matrix.txt

```

<br>
<div class="rectangle"><h2 style="color:white">&nbsp;&nbsp;Clustering Statistics</h2></div>
<br>

### Number of ncASVs per clustering percentage -- "1" represents number of ASVs

<br>

```{r nuc_number, echo=FALSE}
nn=read.csv("number_per_percentage_nucl.csv", header=TRUE)
#nn=read.csv("number_per_percentage_nucl.csv", header=T)
nnp <- plot_ly(nn, x=nn[,1], y=nn[,2], type="scatter", mode = 'lines+markers', marker=list(color='#088da5', line=list(color = 'black',
               width = .1)), hovertemplate = paste('ID%: %{x}','<br>Number of ncASVs: %{y}','<extra></extra>'))
nnp <- nnp %>% layout(yaxis = list(title = "Number of ncASVs"))
nnp <- nnp %>% layout(xaxis = list(title = "Clustering ID %"))
nnp <- nnp %>% config(toImageButtonOptions=list(format='svg',filename='NuclClustresults', height= 500, width= 800, scale= 1))
nnp
```
NOTE: The "1" on the x-axis represents number of ASVs identified by vsearch
<br>

<div class="rectangle"><h2 style="color:white">&nbsp;&nbsp;ASV Pairwise Distance Heatmap</h2></div>

<br>
<br>
```{r asvheatmap, echo=FALSE}
simmatrix<- read.csv("asv_matrix.txt", header=FALSE)
rownames(simmatrix) <- simmatrix[,1]
simmatrix <- simmatrix[,-1]
colnames(simmatrix) <-rownames(simmatrix)
cols <- dim(simmatrix)[2]
simmatrix$AA <- rownames(simmatrix)
rval=nrow(simmatrix)
simmatrix2 <- simmatrix %>%
  gather(1:rval, key=sequence, value=similarity)
x=reorder(simmatrix2$AA,simmatrix2$similarity)
y=reorder(simmatrix2$sequence,simmatrix2$similarity)
similaritymatrix <- ggplot(simmatrix2, aes(x=x, y=y,fill=similarity))+
      geom_raster()+
      scale_fill_distiller(palette="Spectral")+
      theme(axis.text.x = element_text(angle = 90))+
      theme(axis.title.x=element_blank())+
      theme(axis.title.y=element_blank())

heat <- ggplotly(similaritymatrix)
heat <- heat %>% config(toImageButtonOptions=list(format='svg',filename='heatmap', height= 500, width= 800, scale= 1))
heat
```
<br>

### ASV Shannon Entropy Analysis (https://merenlab.org/2012/05/11/oligotyping-pipeline-explained/)

<br>

```{r med_asv, echo=FALSE}
if (params$asvMED == "true") {
    med_asv=read.delim(paste(params$projtag, sep="","_ASVs_Aligned_informativeonly.fasta-ENTROPY"), sep = ",", header=TRUE)

    mplot=plot_ly(med_asv, x=med_asv$Base_position, y=med_asv$Shannons_Entropy, type="bar", name="Entropy", opacity = .8, line=list(color='#088da5'),
                  hovertemplate = paste('Base position: %{x}','<br>Entropy: %{y}','<extra></extra>'))
    mplot<- mplot %>% layout(xaxis=list(title="Base position"))
    mplot<- mplot %>% layout(yaxis=list(title="Entropy"))
    mplot<- mplot %>% config(toImageButtonOptions=list(format='svg',filename='ASV_MED', height= 500, width= 800, scale= 1), displaylogo = FALSE)
    mplot
} else {
    print("ASV Shannon Entropy analyses not signaled to run. To run, add the options --asvMED and -resume to your DataCheck launch command.")
}
```
<br>

### ASV Shannon Entropy Peaks Breakdown

<br>

```{r med_asv_peak, echo=FALSE}
if (params$asvMED == "true") {
med_asv_csv=read.csv(paste(params$projtag, sep="", "_ASV_entropy_breakdown.csv"), header=TRUE)
paged_table(med_asv_csv, options = list(rows.print = 10))
} else {
    print("ASV Shannon Entropy analyses not signaled to run. To run, add the options --asvMED and -resume to your DataCheck launch command.")
}
```
<br>

### Number of pcASVs per clustering percentage

<br>

```{r prot_number, echo=FALSE}
pn=read.csv("number_per_percentage_prot.csv", header=TRUE)
#pn=read.csv("number_per_percentage_prot.csv", header=T)
pnp <- plot_ly(pn, x=pn[,1], y=pn[,2], type="scatter", mode = 'lines+markers', marker=list(color='#088da5', line=list(color = 'black',
               width = .1)), hovertemplate = paste('ID%: %{x}','<br>Number of pcASVs: %{y}','<extra></extra>'))
pnp <- pnp %>% layout(yaxis = list(title = "Number of pcASVs"))
pnp <- pnp %>% layout(xaxis = list(title = "Clustering ID %"))
pnp <- pnp %>% config(toImageButtonOptions=list(format='svg',filename='Protclustresults', height= 500, width= 800, scale= 1))
pnp
```
NOTE: The "1" represents the number of AminoTypes which are unique amino acid sequences (AminoTypes) in your dataset

<br>
<div class="rectangle"><h2 style="color:white">&nbsp;&nbsp;AminoType Pairwise Distance Heatmap</h2></div>

<br>
<br>
```{r aminoheatmap, echo=FALSE}
simmatrix<- read.csv("amino_matrix.txt", header=FALSE)
rownames(simmatrix) <- simmatrix[,1]
simmatrix <- simmatrix[,-1]
colnames(simmatrix) <-rownames(simmatrix)
cols <- dim(simmatrix)[2]
simmatrix$AA <- rownames(simmatrix)
rval=nrow(simmatrix)
simmatrix2 <- simmatrix %>%
  gather(1:rval, key=sequence, value=similarity)
x=reorder(simmatrix2$AA,simmatrix2$similarity)
y=reorder(simmatrix2$sequence,simmatrix2$similarity)
similaritymatrix <- ggplot(simmatrix2, aes(x=x, y=y,fill=similarity))+
      geom_raster()+
      scale_fill_distiller(palette="Spectral")+
      theme(axis.text.x = element_text(angle = 90))+
      theme(axis.title.x=element_blank())+
      theme(axis.title.y=element_blank())

heat <- ggplotly(similaritymatrix)
heat <- heat %>% config(toImageButtonOptions=list(format='svg',filename='heatmap', height= 500, width= 800, scale= 1))
heat
```
<br>
<br>

### AminoTypes Shannon Entropy Analysis (https://merenlab.org/2012/05/11/oligotyping-pipeline-explained/)

<br>

```{r med_amino, echo=FALSE}
if (params$aminoMED == "true") {
med_amino=read.delim(paste(params$projtag, sep="","_AminoTypes_Aligned_informativeonly.fasta-ENTROPY"), sep = ",", header=TRUE)

mplot=plot_ly(med_amino, x=med_amino$Base_position, y=med_amino$Shannons_Entropy, type="bar", name="Entropy", opacity = .8, line=list(color='#088da5'),
              hovertemplate = paste('Base position: %{x}','<br>Entropy: %{y}','<extra></extra>'))
mplot<- mplot %>% layout(xaxis=list(title="Base position"))
mplot<- mplot %>% layout(yaxis=list(title="Entropy"))
mplot<- mplot %>% config(toImageButtonOptions=list(format='svg',filename='AminoTypes_MED', height= 500, width= 800, scale= 1), displaylogo = FALSE)
mplot
} else {
    print("AminoType Shannon Entropy analyses not signaled to run. To run, add the options --aminoMED and -resume to your DataCheck launch command.")
}
```
<br>

### AminoTypes Shannon Entropy Peaks Breakdown

<br>

```{r med_amino_peak, echo=FALSE}
if (params$aminoMED == "true") {
med_amino_csv=read.csv(paste(params$projtag, sep="", "_AminoType_entropy_breakdown.csv"), header=TRUE)
paged_table(med_amino_csv, options = list(rows.print = 10))
} else {
    print("AminoType Shannon Entropy analyses not signaled to run. To run, add the options --aminoMED and -resume to your DataCheck launch command.")
}
```
<br>
<br>
<br>
<br>
<br>
<br>
<br>
<br>
<br>
<br>
<br>
<br>
<br>
