---
title: "vAMPirus Analyze Report `r commandArgs(trailingOnly=T)[1]`"
date: "Generated on: `r Sys.time()`"
output: html_document
params:
    interactive: TRUE
    projtag: !r commandArgs(trailingOnly=T)[1]
    skipReadProcessing: !r commandArgs(trailingOnly=T)[2]
    skipMerging: !r commandArgs(trailingOnly=T)[3]
    skipAdapterRemoval: !r commandArgs(trailingOnly=T)[4]
    skipTaxonomy: !r commandArgs(trailingOnly=T)[5]
    skipPhylogeny: !r commandArgs(trailingOnly=T)[6]
    trymax: !r commandArgs(trailingOnly=T)[7]
    stats: !r commandArgs(trailingOnly=T)[8]
    metadata: !r commandArgs(trailingOnly=T)[9]
    minimumCounts: !r commandArgs(trailingOnly=T)[10]
    asvMED: !r commandArgs(trailingOnly=T)[11]
    aminoMED: !r commandArgs(trailingOnly=T)[12]
    type: !r commandArgs(trailingOnly=T)[13]
    nodeCol: !r commandArgs(trailingOnly=T)[14]
    asvTClust: !r commandArgs(trailingOnly=T)[15]
    aminoTClust: !r commandArgs(trailingOnly=T)[16]
---
<style>
.rectangle {
  height: 37px;
  width: 100%;
  background-color: #576675;
}
</style>

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE,
      message = FALSE,
      warning = FALSE,
      out.width="100%")
```

```{r pathways, echo=FALSE}
knitr::include_graphics("vamplogo.png")
```

```{r load_libraries, include=FALSE}
library(vegan)
library(tidyverse)
library(scales)
library(cowplot)
library(dplyr)
library(ggtree)
library(plotly)
library(knitr)
library(kableExtra)
library(rmarkdown)
library(processx)
library(ape)
```

```{r colors, include=FALSE}
mycol=c('#088da5','#73cdc8','#ff6f61','#7cb8df','#88b04b','#00a199','#6B5B95','#92A8D1','#b0e0e6','#ff7f50','#088d9b','#E15D44','#e19336')
```
<br>
------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
NOTE: Most plots are interactive and you can use the legend to specify samples/treatment of interest. You can also download an .svg version of each figure within this report.
------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
<br>
<div class="rectangle"><h2 style="color:white">&nbsp;&nbsp;Pre- and Post-Adapter Removal Read Stats</h2></div>
<br>
```{r readstats, echo=FALSE}
if (params$skipReadProcessing == "true" || params$skipMerging == "true" ) {
    writeLines("\n--------------------------------------------------------------\n")
    cat(readLines(list.files(pattern="filter_reads.txt")) , sep = '\n')
    writeLines("\n--------------------------------------------------------------\n")
} else {
    if (params$skipAdapterRemoval == "false") {
        reads_stats=read.csv("final_reads_stats.csv")
        #paged_table(reads_stats,options = list(rows.print = 20))
        knitr::kable(reads_stats, digits = 2, align = 'c', booktabs = TRUE, caption = "Table 1: Read summary stats") %>%
        kable_styling(font_size = 12, full_width = F)%>%
        scroll_box(width = "100%", height = "100%")
    } else {
        writeLines("\n--------------------------------------------------------------\n")
        cat(readLines(list.files(pattern="filter_reads.txt")) , sep = '\n')
        writeLines("\n--------------------------------------------------------------\n")
    }
}
```
<br>
<br>


### Total number of reads before and after adapter removal

```{r readstats_plot, echo=FALSE}
# Plot of reads before and after
if (params$skipReadProcessing == "true" || params$skipMerging == "true" ) {
    writeLines("\n--------------------------------------------------------------\n")
    cat(readLines(list.files(pattern="filter_reads.txt")) , sep = '\n')
    writeLines("\n--------------------------------------------------------------\n")
} else {
    if (params$skipAdapterRemoval == "false") {
        ptotal <- plot_ly(type="box",marker=list(colors=mycol))
        ptotal <- ptotal %>% add_boxplot(y=reads_stats$Total_before, name="Reads before filtering")
        ptotal <- ptotal %>% add_boxplot(y=reads_stats$Total_after, name="Reads after filtering")
        #ptotal <- ptotal %>% layout(title=list(text="Number of reads before and after filtering"))
        ptotal <- ptotal %>% layout(legend = list(x=10,y=.5))
        ptotal <- ptotal %>% config(toImageButtonOptions=list(format='svg',filename='TotReads_b4_af_adaptrem', height= 500, width= 800, scale= 1))
        ptotal
    } else {
        # file fore here
        writeLines("\n--------------------------------------------------------------\n")
        cat(readLines(list.files(pattern="filter_reads.txt")) , sep = '\n')
        writeLines("\n--------------------------------------------------------------\n")
    }
}
```
<br>

### Forward (R1) and reverse (R2) read length before and after adapter removal

```{r readstats_plot2, echo=FALSE}
# Plot of R1 and R2 before and after
if (params$skipReadProcessing == "true" || params$skipMerging == "true" ) {
    writeLines("\n--------------------------------------------------------------\n")
    cat(readLines(list.files(pattern="filter_reads.txt")) , sep = '\n')
    writeLines("\n--------------------------------------------------------------\n")
} else {
    if (params$skipAdapterRemoval == "false") {
        pr <- plot_ly(y=reads_stats$R1_before_length, type="box", name="R1 length before")
        pr <- pr %>% add_boxplot(y=reads_stats$R1_after_length, name="R1 length after")
        pr <- pr %>% add_boxplot(y=reads_stats$R2_before_length, name="R2 length before")
        pr <- pr %>% add_boxplot(y=reads_stats$R2_after_length, name="R2 length after")
        #pr <- pr %>% layout(title = "R1 and R2 Length")
        pr <- pr %>% layout(legend = list(x=10,y=.5))
        pr <- pr %>% config(toImageButtonOptions=list(format='svg',filename='readlen_b4_af_adaptrem', height= 500, width= 800, scale= 1))
        pr
    } else {
        # file fore here
        writeLines("\n--------------------------------------------------------------\n")
        cat(readLines(list.files(pattern="filter_reads.txt")) , sep = '\n')
        writeLines("\n--------------------------------------------------------------\n")
    }
}
```
<br>
<br>
<br>
```{bash load_datasets_bash, include=FALSE}
if [ `ls *_ASV_Grouping.csv | wc -l` -ge 1 ];then
  cat *_ASV_Grouping.csv >asv_medfile.csv
  mv *_ASV_Groupingcounts.csv asv_groupcounts.csv
  cat asv_groupcounts.csv | awk -F "," '{print $1","$2}' >tree_group.csv
  mv *_ASV_Group_Reps_iq.treefile grouptree.txt
elif [ `ls *_AminoType_Grouping.csv | wc -l` -ge 1 ];then
  cat *_AminoType_Grouping.csv >amino_medfile.csv
  mv *_AminoType_Groupingcounts.csv amino_groupcounts.csv
  cat amino_groupcounts.csv | awk -F "," '{print $1","$2}' >tree_group.csv
  mv *_AminoType_Group_Reps_iq.treefile grouptree.txt
fi
cat *_counts.csv >counts.csv
cat *_PercentID.matrix >matrix.txt
if [ `ls -1 *_PercentID.matrix | wc -l` -ge 1 ];then
  cat *_summary_for_plot.csv >sum.csv
else
  echo "Taxonomy analysis was skipped" >tax.txt
fi
if [ `ls -1 *_iq.treefile | wc -l` -ge 1 ];then
  cat *_iq.treefile >tree.txt
else
  echo "Phylogeny analysis was skipped" >tree.txt
fi
if [ `ls -1 *_quicker_taxbreakdown.csv | wc -l` -ge 1 ];then
    cat *_quicker_taxbreakdown.csv > quicker_taxbreakdown.csv
fi
```
```{r load_datasets, include=FALSE}
sample_name="counts.csv"
sample_metadata=params$metadata
#sample_metadata="rna_virus_meta.csv"
data<- read.csv(sample_name, check.names=FALSE)
data2 <-as.data.frame(t(data))
data2$sample <- row.names(data2)
colnames(data2)<- as.matrix(data2[1,])
as.data.frame(data2)
data2 <- data2[-1,]

#X.OTU.ID for X.Sequence.
data2 <- data2 %>%
  rename(sample=OTU_ID)
data2dim <- dim(data2)

##Loading metadata
samples <- read.csv(sample_metadata, header = TRUE)

##Combining data and metadata
data3 <- merge(data2, samples, by="sample")

dim_data3 <- dim(data3)
dim_samples <- dim(samples)
cols <- dim_data3[2]-dim_samples[2]+1
first <-colnames(data3)[2]
last <- colnames(data3)[cols]
data3[,2:cols] <- lapply(data3[,2:cols], as.character)
data3[,2:cols] <- lapply(data3[,2:cols], as.numeric)

#Calculate total reads per sample
data4 <- data3%>%
  mutate(sum=select(.,2:cols)%>%
           apply(1, sum, na.rm=TRUE))

```


<div class="rectangle"><h2 style="color:white">&nbsp;&nbsp;Number of Reads Per Sample</h2></div>

```{r plot, echo=FALSE}
# sample and count
con <- plot_ly(data4, x = ~sum, y = ~sample, name = "Sample", type = 'scatter',
               mode = "markers", marker = list(color = "#088da5"), hovertemplate = paste('Sample: %{y}','<br>Total reads: %{x}','<extra></extra>'))
con <- con %>% layout(xaxis = list(title = "Total reads"),yaxis = list(title = "Sample"))
con <- con %>% config(toImageButtonOptions=list(format='svg',filename='Counts_per_sample', height= 500, width= 800, scale= 1))
con
```
<br>
<br>
<br>
```{r filter_data, include=FALSE}
##Filter samples with low reads
nfil=params$minimumCounts
#nfil=1000
data5 <- data4 %>%
  filter(sum>nfil)
   #can cause errors
data5dim <-dim(data5)
minreads<-min(data5$sum)

```
<br>
<br>
<br>
<br>
<br>

<div class="rectangle"><h2 style="color:white">&nbsp;&nbsp;Rarefaction Curves</h2></div>
```{r rarefaction, echo=FALSE, cache=FALSE}
##Rarefaction curves
rarefaction <- rarecurve(data5[,2:cols])

##rarefied dataset
raredata <- as.data.frame(rrarefy(data5[,2:cols], sample=minreads))
```
<br>
<br>
<br>
<br>
<br>


<div class="rectangle"><h2 style="color:white">&nbsp;&nbsp;Diversity Analyses Plots</h2></div>

<br>
<br>

### Shannon diversity

<br>
```{r diversity_analysis, echo=FALSE}
metadata <- data5[,(cols+1):data5dim[2]]
metadata$sample <- data5$sample
index <-diversity(raredata, index= "shannon")
shannondata5 <- as.data.frame(index)
shannondata5$sample<- data5$sample
shannondata5_2 <- merge(shannondata5, metadata, by="sample")

sh <- plot_ly(shannondata5_2, x=~treatment, y=~index, color=~treatment, colors=mycol, type="box", boxpoints = "all", pointpos = 0, jitter = 0.5)
#sh <- sh %>% layout(title = list(text="Shannon diversty",y=.99))
sh <- sh %>% layout(legend = list(x=10,y=.5), yaxis=list(title = "Index"), xaxis=list(title = "Treatment"))
sh <- sh %>% config(toImageButtonOptions=list(format='svg',filename='ShannonDiv', height= 500, width= 800, scale= 1))
sh

if (params$stats == "true" ) {
    shannonaov <- aov(index ~ treatment, data= shannondata5_2)
    st <- shapiro.test(resid(shannonaov))
    bt <- bartlett.test(index ~ treatment, data= shannondata5_2)

    if (st$p.value > .05 && bt$p.value > .05) {
      print("Shapiro Test of normality - data is normal p-value > 0.05")
      print(shapiro.test(resid(shannonaov)))
      writeLines("\n--------------------------------------------------------------\n")
      print("Bartlett Test variance homogeneity - variance is homogeneous p-value > 0.05")
      print(bartlett.test(index ~ treatment, data= shannondata5_2))
      writeLines("\n--------------------------------------------------------------\n")
      print("ANOVA Results")
      print(summary(shannonaov))
      writeLines("\n--------------------------------------------------------------\n")
      #Tukey Honest Significant Differences (pairwise comparison) - significant p <.05
      print("Tukey HSD - Pairwise comparison - significant differences indicated by p-value < 0.05")
      print(TukeyHSD(shannonaov))
      writeLines("\n--------------------------------------------------------------\n")
    } else {
      print("Shapiro Test of normality - data is normal if p-value > 0.05")
      print(shapiro.test(resid(shannonaov)))
      writeLines("\n--------------------------------------------------------------\n")
      print("Bartlett Test variance homogeneity - variance is homogeneous if p-value > 0.05")
      print(bartlett.test(index ~ treatment, data= shannondata5_2))
      writeLines("\n--------------------------------------------------------------\n")
      print("Data either not normal or variance not homogenous")
      print("Kruskal-Wallis Test - test significant if p <.05")
      #Kruskal-Wallis test - significant p <.05
      mykt <- kruskal.test(index ~ treatment, data= shannondata5_2)
      print(mykt)
      writeLines("\n--------------------------------------------------------------\n")
      if (mykt$p.value < .05) {
        #Pairwise comparison
        print("Wilcox.test - pairwise comparison")
        print(pairwise.wilcox.test(shannondata5_2$index, shannondata5_2$treatment, p.adjust.method = "BH"))
      } else {
        print("Data not significant. Skipping pairwise comparison")
      }
      writeLines("\n--------------------------------------------------------------\n")
      print("ANOVA - one or more of the assumptions not met, take with a grain of salt.")
      print(summary(shannonaov))
      writeLines("\n--------------------------------------------------------------\n")
    }
} else {
    print("Stats skipped. To toggle on use \"--stats run\" in vAMPirus launch command")
}
```
<br>
<br>
<br>
<br>

### Simpson diversity

<br>
```{r diversity_analysis2, echo=FALSE}
index <- diversity(raredata, index= "simpson")
simpsondata5 <- as.data.frame(index)
simpsondata5$sample<- data5$sample
simpsondata5_2 <- merge(simpsondata5, metadata, by="sample")

s <- plot_ly(simpsondata5_2, x=~treatment, y=~index, color=~treatment, colors=mycol, type="box", boxpoints = "all", pointpos = 0, jitter = 0.5)
#s <- s %>% layout(title = list(text="Simpson diversty",y=.99))
s <- s %>% layout(legend = list(x=10,y=.5), yaxis=list(title = "Index"), xaxis=list(title = "Treatment"))
s <- s %>% config(toImageButtonOptions=list(format='svg',filename='SimpsonDiv', height= 500, width= 800, scale= 1))
s

if (params$stats == "true" ) {
    simpsonaov <- aov(index ~ treatment, data= simpsondata5_2)
    st <- shapiro.test(resid(simpsonaov))
    bt <- bartlett.test(index ~ treatment, data= simpsondata5_2)

    if (st$p.value > .05 && bt$p.value > .05) {
      print("Shapiro Test of normality - data is normal p-value > 0.05")
      print(shapiro.test(resid(simpsonaov)))
      writeLines("\n--------------------------------------------------------------\n")
      print("Bartlett Test variance homogeneity - variance is homogeneous p-value > 0.05")
      print(bartlett.test(index ~ treatment, data= simpsondata5_2))
      writeLines("\n--------------------------------------------------------------\n")
      print("ANOVA Results")
      print(summary(simpsonaov))
      writeLines("\n--------------------------------------------------------------\n")
      #Tukey Honest Significant Differences (pairwise comparison) - significant p <.05
      print("Tukey HSD - Pairwise comparison - significant differences indicated by p-value < 0.05")
      print(TukeyHSD(simpsonaov))
      writeLines("\n--------------------------------------------------------------\n")
    } else {
      print("Shapiro Test of normality - data is normal if p-value > 0.05")
      print(shapiro.test(resid(simpsonaov)))
      writeLines("\n--------------------------------------------------------------\n")
      print("Bartlett Test variance homogeneity - variance is homogeneous if p-value > 0.05")
      print(bartlett.test(index ~ treatment, data= simpsondata5_2))
      writeLines("\n--------------------------------------------------------------\n")
      print("Data either not normal or variance not homogenous")
      print("Kruskal-Wallis Test - test significant if p <.05")
      #Kruskal-Wallis test - significant p <.05
      mykt <- kruskal.test(index ~ treatment, data= simpsondata5_2)
      print(mykt)
      writeLines("\n--------------------------------------------------------------\n")
      if (mykt$p.value < .05) {
        #Pairwise comparison
        print("Wilcox.test - pairwise comparison")
        print(pairwise.wilcox.test(simpsondata5_2$index, simpsondata5_2$treatment, p.adjust.method = "BH"))
      } else {
        print("Data not significant. Skipping pairwise comparison")
      }
      writeLines("\n--------------------------------------------------------------\n")
      print("ANOVA - one or more of the assumptions not met, take with a grain of salt.")
      print(summary(simpsonaov))
      writeLines("\n--------------------------------------------------------------\n")
    }
} else {
    print("Stats skipped. To toggle on use \"--stats run\" in vAMPirus launch command")
}
```
<br>
<br>
<br>
<br>

### Richness

<br>
```{r diversity_analysis3, echo=FALSE}
mind5<-min(data5$sum)
index <- rarefy(data5[,2:cols], sample=mind5)
rarerichnessdata5 <- as.data.frame(index)
rarerichnessdata5$sample <-data5$sample
richdata5_2 <- merge(rarerichnessdata5, metadata, by="sample")

ri <- plot_ly(richdata5_2, x=~treatment, y=~index, color=~treatment, colors=mycol, type="box", boxpoints = "all", pointpos = 0, jitter = 0.5)
#ri <- ri %>% layout(title = list(text="ASV Richness",y=.99))
ri <- ri %>% layout(legend = list(x=10,y=.5), yaxis=list(title = "Index"), xaxis=list(title = "Treatment"))
ri <- ri %>% config(toImageButtonOptions=list(format='svg',filename='SpeciesRich', height= 500, width= 800, scale= 1))
ri

if (params$stats == "true" ) {
    richaov <- aov(index ~ treatment, data= richdata5_2)
    st <- shapiro.test(resid(richaov))
    bt <- bartlett.test(index ~ treatment, data= richdata5_2)

    if (st$p.value > .05 && bt$p.value > .05) {
      print("Shapiro Test of normality - data is normal p-value > 0.05")
      print(shapiro.test(resid(richaov)))
      writeLines("\n--------------------------------------------------------------\n")
      print("Bartlett Test variance homogeneity - variance is homogeneous p-value > 0.05")
      print(bartlett.test(index ~ treatment, data= richdata5_2))
      writeLines("\n--------------------------------------------------------------\n")
      print("ANOVA Results")
      print(summary(richaov))
      #Tukey Honest Significant Differences (pairwise comparison) - significant p <.05
      writeLines("\n--------------------------------------------------------------\n")
      print("Tukey HSD - Pairwise comparison - significant differences indicated by p-value < 0.05")
      print(TukeyHSD(richaov))
      writeLines("\n--------------------------------------------------------------\n")
    } else {
      print("Shapiro Test of normality - data is normal if p-value > 0.05")
      print(shapiro.test(resid(richaov)))
      writeLines("\n--------------------------------------------------------------\n")
      print("Bartlett Test variance homogeneity - variance is homogeneous if p-value > 0.05")
      print(bartlett.test(index ~ treatment, data= richdata5_2))
      writeLines("\n--------------------------------------------------------------\n")
      print("Data either not normal or variance not homogenous")
      print("Kruskal-Wallis Test - test significant if p <.05")
      #Kruskal-Wallis test - significant p <.05
      mykt <- kruskal.test(index ~ treatment, data= richdata5_2)
      print(mykt)
      writeLines("\n--------------------------------------------------------------\n")
      if (mykt$p.value < .05) {
        #Pairwise comparison
        print("Wilcox.test - pairwise comparison")
        print(pairwise.wilcox.test(richdata5_2$index, richdata5_2$treatment, p.adjust.method = "BH"))
      } else {
        print("Data not significant. Skipping pairwise comparison")
      }
      writeLines("\n--------------------------------------------------------------\n")
      print("ANOVA - one or more of the assumptions not met, take with a grain of salt.")
      print(summary(richaov))
      writeLines("\n--------------------------------------------------------------\n")
    }
} else {
    print("Stats skipped. To toggle on use \"--stats run\" in vAMPirus launch command")
}
```
<br>
<br>
<br>
<br>
<br>

<div class="rectangle"><h2 style="color:white">&nbsp;&nbsp;Distance To Centroid</h2></div>
<br>
```{r distance, echo=FALSE}
##Distance
intermediate <- raredata
bray.distance <- vegdist(sqrt(intermediate), method="bray")

##Dispersion
disper <- betadisper(bray.distance, group = metadata$treatment, type="centroid")
df <- data.frame(Distance_to_centroid=disper$distances,Group=disper$group)
df$sample <- data5$sample
df2 <- merge(df, metadata, by="sample")

cen <- plot_ly(df2, x=~treatment, y=~Distance_to_centroid, color=~treatment, colors=mycol, type="box", boxpoints = "all", pointpos = 0, jitter = 0.5)
#cen <- cen %>% layout(title = list(text="Distance to centroid",y=.99))
cen <- cen %>% layout(legend = list(x=10,y=.5), yaxis=list(title = "Distance"), xaxis=list(title = "Treatment"))
cen <- cen %>% config(toImageButtonOptions=list(format='svg',filename='Dispersion', height= 500, width= 800, scale= 1))
cen

if (params$stats == "true" ) {
    adn <- adonis(bray.distance~data5$treatment)
    adn
} else {
    print("Stats skipped. To toggle on use \"--stats run\" in vAMPirus launch command")
}
```
<br>
<br>
<br>
<br>
<br>

<div class="rectangle"><h2 style="color:white">&nbsp;&nbsp;NMDS Plots</h2></div>

<br>

### 2D NMDS

<br>
```{r nmds2d, echo=FALSE}
##NMDS
datax <- decostand(raredata, method="total") #method 'total' normalizes data to sum up to 1 --data5[,2:cols]

MDS <- metaMDS(sqrt(datax),
               distance = "bray",autotransform = FALSE,
               k = 2,
               maxit = 999,
               trymax = params$try,
               wascores = TRUE)

if (MDS$converged == "TRUE") {

data.scores <- as.data.frame(scores(MDS))
data.scores$sample <- data5$sample
data.scores.2 <- merge(data.scores, metadata, by="sample")

p <- ggplot(data.scores.2, aes(x=NMDS1, y=NMDS2,color=treatment))+
  geom_point(size=2)+
  theme_classic()+
  theme(legend.title = element_blank())

fff <-plot_ly(data.scores.2, x=~NMDS1, y=~NMDS2, color=~treatment, colors=mycol, text = ~paste("Sample: ", sample))
fff <- fff %>% layout(legend=list(y=.5))
fff <- fff %>% config(toImageButtonOptions=list(format='svg',filename='2Dnmds', height= 500, width= 800, scale= 1))
fff

} else {
print("NMDS did not converge. Printing PCoA")
#calculate bray curtis distance
bcdist <- vegdist(datax, "bray")
res <- pcoa(bcdist)
#res$values
#biplot(res) #to make the boring 2D PCoA
comp <- as.data.frame(res$vectors)
comp$sample <- data5$sample
comp2 <- merge(comp, metadata, by="sample")
fig <- plot_ly(comp2, x = ~Axis.1, y = ~Axis.2, color= ~treatment, text = ~paste("Sample: ", sample,"<br>Treatment: ",treatment))
fig <- fig %>% layout(xaxis = list(title = "PCoA 1"),
                      yaxis = list(title = "PCoA 2"))
fig <- fig %>% layout(legend=list(y=.5))
fig <- fig %>% config(toImageButtonOptions=list(format='svg',filename='2D_PCoA', height= 500, width= 800, scale= 1))
fig
}
```

<br>
<br>
<br>
<br>

### 3D NMDS

<br>

```{r nmds3d, echo=FALSE}
MDS3 <- metaMDS(sqrt(datax),
               distance = "bray",autotransform = FALSE,
               k = 3,
               maxit = 999,
               trymax = params$try,
               wascores = TRUE)

if (MDS3$converged == "TRUE") {

data.scores3 <- as.data.frame(scores(MDS3))
data.scores3$sample <- data5$sample
data.scores.3 <- merge(data.scores3, metadata, by="sample")
p3d <- plot_ly(data.scores.3, x=~NMDS1, y=~NMDS2, z=~NMDS3, text=~paste("Sample: ", sample),
        color=~treatment, colors=mycol,
        mode = 'markers', symbol = ~treatment, symbols = c('square','circle'),
        marker = list(opacity = .8,line=list(color = 'darkblue',width = 1))
        )
p3d <- p3d %>% layout(legend=list(y=.5))
p3d <- p3d %>% config(toImageButtonOptions=list(format='svg',filename='3Dnmds', height= 500, width= 800, scale= 1))
p3d

} else {
print("NMDS did not converge. Printing PCoA")
#calculate bray curtis distance
bcdist <- vegdist(datax, "bray")
res <- pcoa(bcdist)
comp <- as.data.frame(res$vectors)
comp$sample <- data5$sample
comp2 <- merge(comp, metadata, by="sample")
fig <- plot_ly(comp2, x = ~Axis.1, y = ~Axis.2, z = ~Axis.3, color= ~treatment, text = ~paste("Sample: ", sample,"<br>Treatment: ",treatment))
fig <- fig %>% layout(xaxis = list(title = "PCoA 1"),
                      yaxis = list(title = "PCoA 2"),
                      zaxis = list(title = "PCoA 3"))
fig <- fig %>% layout(legend=list(y=.5))
fig <- fig %>% config(toImageButtonOptions=list(format='svg',filename='3D_PCoA', height= 500, width= 800, scale= 1))
fig
}
```


<br>
<br>
<br>
<br>
<br>

<div class="rectangle"><h2 style="color:white">&nbsp;&nbsp;Relative Abundance Per Sample</h2></div>

```{r long, echo=FALSE}
dataz <- decostand(data5[,2:cols],method="total") #method 'total' normalizes data to sum up to 1
dataz$sample <- data5$sample
datay <- merge(dataz, metadata, by="sample")
datalong <- datay %>%
  tidyr::gather(first:last, key=hit, value=reads)

ddd <- plot_ly(datalong, x=~sample, y=~reads, color=~hit, colors=mycol)
ddd <- ddd %>% layout(type='bar',barmode = 'stack')
ddd <- ddd %>% layout(legend = list(x=10,y=.5), xaxis=list(title = "Sample"), yaxis=list(title = "Relative abundance"))
ddd <- ddd %>% config(toImageButtonOptions=list(format='svg',filename='Relative_abundance', height= 500, width= 800, scale= 1))
ddd
```


<br>
<br>
<br>
<br>
<br>

<div class="rectangle"><h2 style="color:white">&nbsp;&nbsp;Sequence Abundance Per Treatment</h2></div>

```{r asv_barplot, echo=FALSE}
datalong <- datalong %>%
  filter(reads>0)

asp2 <- plot_ly(datalong, y=~hit, x=~reads, color=~treatment, colors=mycol, text = ~paste("Sample: ", sample), opacity=.9)
asp2 <- asp2 %>% layout(type='bar', barmode = 'group')
asp2 <- asp2 %>% layout(yaxis = list(title = '', categoryorder = "total ascending"), legend = list(x=10,y=.5))
asp2 <- asp2 %>% config(toImageButtonOptions=list(format='svg',filename='Most_abundant_hits_per_sample', height= 500, width= 800, scale= 1))
asp2
```

<br>
<br>
<br>
<br>
<br>

<div class="rectangle"><h2 style="color:white">&nbsp;&nbsp;Pairwise Percent Similarity Heatmap</h2></div>

<br>
<br>
```{r heatmap, echo=FALSE}
simmatrix<- read.csv("matrix.txt", header=FALSE)
rownames(simmatrix) <- simmatrix[,1]
simmatrix <- simmatrix[,-1]
colnames(simmatrix) <-rownames(simmatrix)
cols <- dim(simmatrix)[2]
simmatrix$AA <- rownames(simmatrix)
rval=nrow(simmatrix)
simmatrix2 <- simmatrix %>%
  gather(1:rval, key=sequence, value=similarity)
x=reorder(simmatrix2$AA,simmatrix2$similarity)
y=reorder(simmatrix2$sequence,simmatrix2$similarity)
similaritymatrix <- ggplot(simmatrix2, aes(x=x, y=y,fill=similarity))+
      geom_raster()+
      scale_fill_distiller(palette="Spectral")+
      theme(axis.text.x = element_text(angle = 90))+
      theme(axis.title.x=element_blank())+
      theme(axis.title.y=element_blank())

heat <- ggplotly(similaritymatrix)
heat <- heat %>% config(toImageButtonOptions=list(format='svg',filename='heatmap', height= 500, width= 800, scale= 1))
heat
```
<br>
<br>
<br>
<br>
<br>


<div class="rectangle"><h2 style="color:white">&nbsp;&nbsp;Taxonomy Results Visualization</h2></div>

<br>
<br>

```{r taxonomy, echo=FALSE}
if (params$skipTaxonomy == "true") {
    # file fore here
    writeLines("\n--------------------------------------------------------------\n")
    cat(readLines(list.files(pattern="tax.txt")) , sep = '\n')
    writeLines("\n--------------------------------------------------------------\n")
} else {
    tax=read.csv("sum.csv",header=F)
    tp <- plot_ly(tax, labels = ~V1, values = ~V2)
    tp <- tp %>% add_pie(marker=list(colors=mycol, line=list(color='#000000', width=.5)), hole = 0.6)
    tp <- tp %>% layout(title = "Taxonomy distribution",  showlegend = F,
                        xaxis = list(showgrid = FALSE, zeroline = FALSE, showticklabels = FALSE),
                        yaxis = list(showgrid = FALSE, zeroline = FALSE, showticklabels = FALSE)
                        )
    tp <- tp %>% config(toImageButtonOptions=list(format='svg',filename='TaxDonut', height= 500, width= 800, scale= 1))
    tp
}

```

<br>
<br>
<br>
<br>
<br>

<div class="rectangle"><h2 style="color:white">&nbsp;&nbsp;Phylogenetic Tree</h2></div>

<br>
<br>

```{r tree, echo=FALSE}
if (params$skipPhylogeny == "true") {
    # file fore here
    writeLines("\n--------------------------------------------------------------\n")
    cat(readLines(list.files(pattern="tree.txt")) , sep = '\n')
    writeLines("\n--------------------------------------------------------------\n")
} else if ((params$asvMED == "true" || params$aminoMED == "true") && (params$type == "ASV" || params$type == "AminoType") && (params$nodeCol == "MED")) {
    tree=read.tree("tree.txt")
    p1 <- ggtree(tree)
    id <- tree$tip.label
    dat <- tibble::tibble(id = id)
    metat <- p1$data %>% dplyr::inner_join(dat, c('label' = 'id'))
    treelist=read.csv("tree_group.csv")
    metat2 <- metat %>% dplyr::inner_join(treelist, c('label' = 'OTU_ID'))
    p2 <- p1 + geom_point(data = metat2, aes(x = x, y = y, label = id, color = GroupID))
    gg = ggplotly(p2, tooltip = c("label","GroupID"))
    gg <- gg %>% config(toImageButtonOptions=list(format='svg',filename='Tree_MED_colors', height= 500, width= 800, scale= 1))
    gg
} else if ((params$skipTaxonomy == "false") && (params$nodeCol == "TAX")) {
    taxMap=read.csv("quicker_taxbreakdown.csv", header=T, check.names = F)
    tree=read.tree("tree.txt")
    p1 <- ggtree(tree)
    id <- tree$tip.label
    dat <- tibble::tibble(id = id)
    metat <- p1$data %>% dplyr::inner_join(dat, c('label' = 'id'))
    colnames(taxMap) <- c("label","tax")
    taxMap
    taxDF <- merge(metat, taxMap, by="label")
    p2 <- p1 + geom_point(data = taxDF, aes(x = x, y = y, label = label, color = tax))
    gg = ggplotly(p2, tooltip = c("label","tax"))
    gg <- gg %>% config(toImageButtonOptions=list(format='svg',filename='Tree_Tax_colors', height= 500, width= 800, scale= 1))
    gg
} else {
    #tree=read.newick("vAMPrun_otu.55.raxml.support")
    tree=read.tree("tree.txt")
    p1 <- ggtree(tree)
    id <- tree$tip.label
    dat <- tibble::tibble(id = id)
    metat <- p1$data %>% dplyr::inner_join(dat, c('label' = 'id'))
    p2 <- p1 + geom_point(data = metat, aes(x = x, y = y, label = id, color = id))
    gg = ggplotly(p2, tooltip = "label")
    gg <- gg %>% config(toImageButtonOptions=list(format='svg',filename='Tree_id_colors', height= 500, width= 800, scale= 1))
    gg
}

```
This tree is a maximum likelihood tree made with IQTREE2 and the parameters you specified in the vampirus.config file. Also, this is an interactive tree, you can zoom in and hover on nodes to know the sequence ID. For a better visualization of this tree, you can find the *.treefile with bootstrap support values within the results directory and visualize using programs like FigTree or ITOL. If you ran the MED analysis, the colors of the nodes correspond to the MED group they were assigned to.

<br>
<br>
<br>
# Post Phylogeny-based Clustering Analyses
```{bash bash_phy_table, include=FALSE}
if [[ $(ls | grep -wc "asv_medfile.csv") -eq 1 ]];then
  awk -F "," '{print $2}' asv_medfile.csv | sort | uniq | sort -g >group.list
  for x in $(cat group.list);do
    echo "${x}" >> ${x}.col
    grep -w "$x" asv_medfile.csv | awk -F "," '{print $1}' | sort >> ${x}.col
  done
  paste -d "," *.col > asv_medtable.csv
  rm *.col
else
    echo "Not MED"
fi

if [[ $(ls | grep -wc "amino_medfile.csv") -eq 1 ]];then
  awk -F "," '{print $2}' amino_medfile.csv | sort | awk -F "Group" '{print $2}' | sort -n | uniq > group.list
  for x in $(cat group.list);do
    gr="Group"${x}""
    echo "${gr}" > ${x}.col
    grep -w "$gr" amino_medfile.csv | awk -F "," '{print $1}' | sort >> ${x}.col
  done
  paste -d "," *.col > amino_medtable.csv
  rm *.col
else
      echo "Not MED"
fi
```

```{r PhyloGroup_table, echo=FALSE}
if (params$asvMED == "true" && params$type == "ASV"){
    #med_group=read.csv("asv_medfile.csv", header = F)
    #colnames(med_group) <- c("SequenceID", "Group", "Sequence")
    # change name of sequence to med peak or something
    #paged_table(med_group,options = list(rows.print = 20))
    med_group=read.csv("asv_medtable.csv", header = T)
    knitr::kable(med_group, digits = 2, align = 'c', booktabs = TRUE, caption = "Table: ASV MED Table") %>%
    kable_styling(font_size = 12, full_width = F)%>%
    scroll_box(width = "100%", height = "100%")
}

if (params$aminoMED == "true" && params$type == "AminoType"){
  #med_group=read.csv("amino_medfile.csv", header = F)
  #colnames(med_group) <- c("SequenceID", "Group", "Sequence")
  #paged_table(med_group,options = list(rows.print = 20))
  med_group=read.csv("amino_medtable.csv", header = T)
  knitr::kable(med_group, digits = 2, align = 'c', booktabs = TRUE, caption = "Table: Aminotype MED Table") %>%
  kable_styling(font_size = 12, full_width = F) %>%
  scroll_box(width = "100%", height = "100%")
}
```

<br>
<br>
<br>
<br>
<br>

<div class="rectangle"><h2 style="color:white">&nbsp;&nbsp;PhyloGroup Relative Abundance Plot</h2></div>

<br>
<br>

```{r med_plot, echo=FALSE}
if (params$asvMED == "true" && params$type == "ASV"){
  #medgroup=read.csv("uVP_AminoType_Groupingcounts.csv", check.names=FALSE)
  medgroup=read.csv("asv_groupcounts.csv", check.names=FALSE)
  meddata <- medgroup %>% select(-2)
  meddata2 <- aggregate(.~GroupID, meddata, sum)
  meddata3 <-as.data.frame(t(meddata2))
  meddata3$sample <- row.names(meddata3)
  colnames(meddata3)<- as.matrix(meddata3[1,])
  as.data.frame(meddata3)
  meddata3 <- meddata3[-1,]
  meddata3 <- meddata3 %>%
    rename(sample=GroupID)
  meddata3dim <- dim(meddata3)
  ##Loading metadata
  samples <- read.csv(sample_metadata, header = TRUE)
  #samples <- read.csv("rna_virus_meta.csv", header = TRUE)
  ##Combining data and metadata
  meddata4 <- merge(meddata3, samples, by="sample")
  #if "meddata4" removed it does not work
  #meddata4
  dim_meddata4 <- dim(meddata4)
  dim_samples <- dim(samples)
  cols <- dim_meddata4[2]-dim_samples[2]+1
  first <-colnames(meddata4)[2]
  last <- colnames(meddata4)[cols]
  meddata4[,2:cols] <- lapply(meddata4[,2:cols], as.character)
  meddata4[,2:cols] <- lapply(meddata4[,2:cols], as.numeric)
  #Calculate total reads per sample
  meddata5 <- meddata4%>%
    mutate(sum=select(.,2:cols)%>%
             apply(1, sum, na.rm=TRUE))
  #meddata5
  ##Filter samples with low reads
  nfil=params$minimumCounts
  #nfil=1000
  meddata5 <- meddata5 %>%
    filter(sum>nfil)
     #can cause errors
  meddata5dim <-dim(meddata5)
  minreads<-min(meddata5$sum)

  dataz <- decostand(meddata5[,2:cols],method="total") #method 'total' normalizes data to sum up to 1
  dataz$sample <- meddata5$sample
  metadata <- meddata5[,(cols+1):meddata5dim[2]]
  metadata$sample <- meddata5$sample
  datay <- merge(dataz, metadata, by="sample")
  datalong <- datay %>%
    tidyr::gather(first:last, key=hit, value=reads)
  ddd <- plot_ly(datalong, x=~sample, y=~reads, color=~hit, colors=mycol, type='bar')
  ddd <- ddd %>% layout(barmode = 'stack')
  ddd <- ddd %>% layout(legend = list(x=10,y=.5), xaxis=list(title = "Sample"), yaxis=list(title = "Relative abundance"))
  ddd <- ddd %>% config(toImageButtonOptions=list(format='svg',filename='Relative_abundance', height= 500, width= 800, scale= 1))
  ddd
}
if (params$aminoMED == "true" && params$type == "AminoType"){
  #medgroup=read.csv("uVP_AminoType_Groupingcounts.csv")
  medgroup=read.csv("amino_groupcounts.csv",  check.names=FALSE)
  meddata <- medgroup %>% select(-2)
  meddata2 <- aggregate(.~GroupID, meddata, sum)
  meddata3 <-as.data.frame(t(meddata2))
  meddata3$sample <- row.names(meddata3)
  colnames(meddata3)<- as.matrix(meddata3[1,])
  as.data.frame(meddata3)
  meddata3 <- meddata3[-1,]
  meddata3 <- meddata3 %>%
    rename(sample=GroupID)
  meddata3dim <- dim(meddata3)
  ##Loading metadata
  samples <- read.csv(sample_metadata, header = TRUE)
  ##Combining data and metadata
  meddata4 <- merge(meddata3, samples, by="sample")
  #if "meddata4" removed it does not work
  #meddata4
  dim_meddata4 <- dim(meddata4)
  dim_samples <- dim(samples)
  cols <- dim_meddata4[2]-dim_samples[2]+1
  first <-colnames(meddata4)[2]
  last <- colnames(meddata4)[cols]
  meddata4[,2:cols] <- lapply(meddata4[,2:cols], as.character)
  meddata4[,2:cols] <- lapply(meddata4[,2:cols], as.numeric)
  #Calculate total reads per sample
  meddata5 <- meddata4%>%
    mutate(sum=select(.,2:cols)%>%
             apply(1, sum, na.rm=TRUE))
  #meddata5
  ##Filter samples with low reads
  nfil=params$minimumCounts
  #nfil=1000
  meddata5 <- meddata5 %>%
    filter(sum>nfil)
     #can cause errors
  meddata5dim <-dim(meddata5)
  minreads<-min(meddata5$sum)

  dataz <- decostand(meddata5[,2:cols],method="total") #method 'total' normalizes data to sum up to 1
  dataz$sample <- meddata5$sample
  metadata <- meddata5[,(cols+1):meddata5dim[2]]
  metadata$sample <- meddata5$sample
  datay <- merge(dataz, metadata, by="sample")
  datalong <- datay %>%
    tidyr::gather(first:last, key=hit, value=reads)
  ddd <- plot_ly(datalong, x=~sample, y=~reads, color=~hit, colors=mycol, type='bar')
  ddd <- ddd %>% layout(barmode = 'stack')
  ddd <- ddd %>% layout(legend = list(x=10,y=.5), xaxis=list(title = "Sample"), yaxis=list(title = "Relative abundance"))
  ddd <- ddd %>% config(toImageButtonOptions=list(format='svg',filename='Relative_abundance', height= 500, width= 800, scale= 1))
  ddd
}
```

<br>
<br>
<br>
<br>
<br>

<div class="rectangle"><h2 style="color:white">&nbsp;&nbsp;PhyloGroup Rarefaction Curves</h2></div>

<br>
<br>

```{r phy_rarefaction, echo=FALSE, cache=FALSE}
if ((params$asvMED == "true" || params$aminoMED == "true") && (params$type == "ASV" || params$type == "AminoType")){
##Rarefaction curves
rarefaction <- rarecurve(meddata5[,2:cols])

##rarefied dataset
raredata <- as.data.frame(rrarefy(meddata5[,2:cols], sample=minreads))
}
```

<br>
<br>
<br>
<br>
<br>
<br>

<div class="rectangle"><h2 style="color:white">&nbsp;&nbsp;PhyloGroup Diversity Analyses Plots</h2></div>

<br>
<br>

### PhyloGroup Shannon diversity

<br>
```{r phy_diversity_analysis1a, echo=FALSE}
if ((params$asvMED == "true" || params$aminoMED == "true") && (params$type == "ASV" || params$type == "AminoType")){
    metadata <- meddata5[,(cols+1):meddata5dim[2]]
    metadata$sample <- meddata5$sample
    index <-diversity(raredata, index= "shannon")
    shannondata5 <- as.data.frame(index)
    shannondata5$sample<- meddata5$sample
    shannondata5_2 <- merge(shannondata5, metadata, by="sample")

    sh <- plot_ly(shannondata5_2, x=~treatment, y=~index, color=~treatment, colors=mycol, type="box", boxpoints = "all", pointpos = 0, jitter = 0.5)
    #sh <- sh %>% layout(title = list(text="Shannon diversty",y=.99))
    sh <- sh %>% layout(legend = list(x=10,y=.5), yaxis=list(title = "Index"), xaxis=list(title = "Treatment"))
    sh <- sh %>% config(toImageButtonOptions=list(format='svg',filename='ShannonDiv', height= 500, width= 800, scale= 1))
    sh
}
```

```{r phy_diversity_analysis1b, echo=FALSE}
#divided because error
if ((params$asvMED == "true" || params$aminoMED == "true") && (params$type == "ASV" || params$type == "AminoType")){
    if (params$stats == "true" ) {
        shannonaov <- aov(index ~ treatment, data= shannondata5_2)
        st <- shapiro.test(resid(shannonaov))
        bt <- bartlett.test(index ~ treatment, data= shannondata5_2)

        if (st$p.value > .05 && bt$p.value > .05) {
          print("Shapiro Test of normality - data is normal p-value > 0.05")
          print(shapiro.test(resid(shannonaov)))
          writeLines("\n--------------------------------------------------------------\n")
          print("Bartlett Test variance homogeneity - variance is homogeneous p-value > 0.05")
          print(bartlett.test(index ~ treatment, data= shannondata5_2))
          writeLines("\n--------------------------------------------------------------\n")
          print("ANOVA Results")
          print(summary(shannonaov))
          writeLines("\n--------------------------------------------------------------\n")
          #Tukey Honest Significant Differences (pairwise comparison) - significant p <.05
          print("Tukey HSD - Pairwise comparison - significant differences indicated by p-value < 0.05")
          print(TukeyHSD(shannonaov))
          writeLines("\n--------------------------------------------------------------\n")
        } else {
          print("Shapiro Test of normality - data is normal if p-value > 0.05")
          print(shapiro.test(resid(shannonaov)))
          writeLines("\n--------------------------------------------------------------\n")
          print("Bartlett Test variance homogeneity - variance is homogeneous if p-value > 0.05")
          print(bartlett.test(index ~ treatment, data= shannondata5_2))
          writeLines("\n--------------------------------------------------------------\n")
          print("Data either not normal or variance not homogenous")
          print("Kruskal-Wallis Test - test significant if p <.05")
          #Kruskal-Wallis test - significant p <.05
          mykt <- kruskal.test(index ~ treatment, data= shannondata5_2)
          print(mykt)
          writeLines("\n--------------------------------------------------------------\n")
          if (mykt$p.value < .05) {
            #Pairwise comparison
            print("Wilcox.test - pairwise comparison")
            print(pairwise.wilcox.test(shannondata5_2$index, shannondata5_2$treatment, p.adjust.method = "BH"))
          } else {
            print("Data not significant. Skipping pairwise comparison")
          }
          writeLines("\n--------------------------------------------------------------\n")
          print("ANOVA - one or more of the assumptions not met, take with a grain of salt.")
          print(summary(shannonaov))
          writeLines("\n--------------------------------------------------------------\n")
        }
    } else {
        print("Stats skipped. To toggle on use \"--stats run\" in vAMPirus launch command")
    }
}
```
<br>
<br>
<br>
<br>

### PhyloGroup Simpson diversity

<br>
```{r phy_diversity_analysis2a, echo=FALSE}
if ((params$asvMED == "true" || params$aminoMED == "true") && (params$type == "ASV" || params$type == "AminoType")){
    index <- diversity(raredata, index= "simpson")
    simpsondata5 <- as.data.frame(index)
    simpsondata5$sample<- meddata5$sample
    simpsondata5_2 <- merge(simpsondata5, metadata, by="sample")

    s <- plot_ly(simpsondata5_2, x=~treatment, y=~index, color=~treatment, colors=mycol, type="box", boxpoints = "all", pointpos = 0, jitter = 0.5)
    #s <- s %>% layout(title = list(text="Simpson diversty",y=.99))
    s <- s %>% layout(legend = list(x=10,y=.5), yaxis=list(title = "Index"), xaxis=list(title = "Treatment"))
    s <- s %>% config(toImageButtonOptions=list(format='svg',filename='SimpsonDiv', height= 500, width= 800, scale= 1))
    s
}
```

```{r phy_diversity_analysis2b, echo=FALSE}
if ((params$asvMED == "true" || params$aminoMED == "true") && (params$type == "ASV" || params$type == "AminoType")){
    if (params$stats == "true" ) {
        simpsonaov <- aov(index ~ treatment, data= simpsondata5_2)
        st <- shapiro.test(resid(simpsonaov))
        bt <- bartlett.test(index ~ treatment, data= simpsondata5_2)

        if (st$p.value > .05 && bt$p.value > .05) {
          print("Shapiro Test of normality - data is normal p-value > 0.05")
          print(shapiro.test(resid(simpsonaov)))
          writeLines("\n--------------------------------------------------------------\n")
          print("Bartlett Test variance homogeneity - variance is homogeneous p-value > 0.05")
          print(bartlett.test(index ~ treatment, data= simpsondata5_2))
          writeLines("\n--------------------------------------------------------------\n")
          print("ANOVA Results")
          print(summary(simpsonaov))
          writeLines("\n--------------------------------------------------------------\n")
          #Tukey Honest Significant Differences (pairwise comparison) - significant p <.05
          print("Tukey HSD - Pairwise comparison - significant differences indicated by p-value < 0.05")
          print(TukeyHSD(simpsonaov))
          writeLines("\n--------------------------------------------------------------\n")
        } else {
          print("Shapiro Test of normality - data is normal if p-value > 0.05")
          print(shapiro.test(resid(simpsonaov)))
          writeLines("\n--------------------------------------------------------------\n")
          print("Bartlett Test variance homogeneity - variance is homogeneous if p-value > 0.05")
          print(bartlett.test(index ~ treatment, data= simpsondata5_2))
          writeLines("\n--------------------------------------------------------------\n")
          print("Data either not normal or variance not homogenous")
          print("Kruskal-Wallis Test - test significant if p <.05")
          #Kruskal-Wallis test - significant p <.05
          mykt <- kruskal.test(index ~ treatment, data= simpsondata5_2)
          print(mykt)
          writeLines("\n--------------------------------------------------------------\n")
          if (mykt$p.value < .05) {
            #Pairwise comparison
            print("Wilcox.test - pairwise comparison")
            print(pairwise.wilcox.test(simpsondata5_2$index, simpsondata5_2$treatment, p.adjust.method = "BH"))
          } else {
            print("Data not significant. Skipping pairwise comparison")
          }
          writeLines("\n--------------------------------------------------------------\n")
          print("ANOVA - one or more of the assumptions not met, take with a grain of salt.")
          print(summary(simpsonaov))
          writeLines("\n--------------------------------------------------------------\n")
        }
    } else {
        print("Stats skipped. To toggle on use \"--stats run\" in vAMPirus launch command")
    }
}
```
<br>
<br>
<br>
<br>

### PhyloGroup Richness

<br>
```{r phy_diversity_analysis3a, echo=FALSE}
if ((params$asvMED == "true" || params$aminoMED == "true") && (params$type == "ASV" || params$type == "AminoType")){
    mind5<-min(meddata5$sum)
    index <- rarefy(meddata5[,2:cols], sample=mind5)
    rarerichnessdata5 <- as.data.frame(index)
    rarerichnessdata5$sample <-meddata5$sample
    richdata5_2 <- merge(rarerichnessdata5, metadata, by="sample")

    ri <- plot_ly(richdata5_2, x=~treatment, y=~index, color=~treatment, colors=mycol, type="box", boxpoints = "all", pointpos = 0, jitter = 0.5)
    #ri <- ri %>% layout(title = list(text="ASV Richness",y=.99))
    ri <- ri %>% layout(legend = list(x=10,y=.5), yaxis=list(title = "Index"), xaxis=list(title = "Treatment"))
    ri <- ri %>% config(toImageButtonOptions=list(format='svg',filename='SpeciesRich', height= 500, width= 800, scale= 1))
    ri
}
```

```{r med_diversity_analysis3b, echo=FALSE}
if ((params$asvMED == "true" || params$aminoMED == "true") && (params$type == "ASV" || params$type == "AminoType")){
    if (params$stats == "true" ) {
        richaov <- aov(index ~ treatment, data= richdata5_2)
        st <- shapiro.test(resid(richaov))
        bt <- bartlett.test(index ~ treatment, data= richdata5_2)

        if (st$p.value > .05 && bt$p.value > .05) {
          print("Shapiro Test of normality - data is normal p-value > 0.05")
          print(shapiro.test(resid(richaov)))
          writeLines("\n--------------------------------------------------------------\n")
          print("Bartlett Test variance homogeneity - variance is homogeneous p-value > 0.05")
          print(bartlett.test(index ~ treatment, data= richdata5_2))
          writeLines("\n--------------------------------------------------------------\n")
          print("ANOVA Results")
          print(summary(richaov))
          #Tukey Honest Significant Differences (pairwise comparison) - significant p <.05
          writeLines("\n--------------------------------------------------------------\n")
          print("Tukey HSD - Pairwise comparison - significant differences indicated by p-value < 0.05")
          print(TukeyHSD(richaov))
          writeLines("\n--------------------------------------------------------------\n")
        } else {
          print("Shapiro Test of normality - data is normal if p-value > 0.05")
          print(shapiro.test(resid(richaov)))
          writeLines("\n--------------------------------------------------------------\n")
          print("Bartlett Test variance homogeneity - variance is homogeneous if p-value > 0.05")
          print(bartlett.test(index ~ treatment, data= richdata5_2))
          writeLines("\n--------------------------------------------------------------\n")
          print("Data either not normal or variance not homogenous")
          print("Kruskal-Wallis Test - test significant if p <.05")
          #Kruskal-Wallis test - significant p <.05
          mykt <- kruskal.test(index ~ treatment, data= richdata5_2)
          print(mykt)
          writeLines("\n--------------------------------------------------------------\n")
          if (mykt$p.value < .05) {
            #Pairwise comparison
            print("Wilcox.test - pairwise comparison")
            print(pairwise.wilcox.test(richdata5_2$index, richdata5_2$treatment, p.adjust.method = "BH"))
          } else {
            print("Data not significant. Skipping pairwise comparison")
          }
          writeLines("\n--------------------------------------------------------------\n")
          print("ANOVA - one or more of the assumptions not met, take with a grain of salt.")
          print(summary(richaov))
          writeLines("\n--------------------------------------------------------------\n")
        }
    } else {
        print("Stats skipped. To toggle on use \"--stats run\" in vAMPirus launch command")
    }
}
```
<br>
<br>
<br>
<br>
<br>

<div class="rectangle"><h2 style="color:white">&nbsp;&nbsp;PhyloGroup Distance To Centroid</h2></div>
<br>
```{r med_distance2a, echo=FALSE}
if ((params$asvMED == "true" || params$aminoMED == "true") && (params$type == "ASV" || params$type == "AminoType")){
    ##Distance
    intermediate <- raredata
    bray.distance <- vegdist(sqrt(intermediate), method="bray")

    ##Dispersion
    disper <- betadisper(bray.distance, group = metadata$treatment, type="centroid")
    df <- data.frame(Distance_to_centroid=disper$distances,Group=disper$group)
    df$sample <- meddata5$sample
    df2 <- merge(df, metadata, by="sample")

    cen <- plot_ly(df2, x=~treatment, y=~Distance_to_centroid, color=~treatment, colors=mycol, type="box", boxpoints = "all", pointpos = 0, jitter = 0.5)
    #cen <- cen %>% layout(title = list(text="Distance to centroid",y=.99))
    cen <- cen %>% layout(legend = list(x=10,y=.5), yaxis=list(title = "Distance"), xaxis=list(title = "Treatment"))
    cen <- cen %>% config(toImageButtonOptions=list(format='svg',filename='Dispersion', height= 500, width= 800, scale= 1))
    cen
}
```
```{r med_distance2b, echo=FALSE}
if ((params$asvMED == "true" || params$aminoMED == "true") && (params$type == "ASV" || params$type == "AminoType")){
    if (params$stats == "true" ) {
        adn <- adonis(bray.distance~meddata5$treatment)
        adn
    } else {
        print("Stats skipped. To toggle on use \"--stats run\" in vAMPirus launch command")
    }
}
```
<br>
<br>
<br>
<br>
<br>

<div class="rectangle"><h2 style="color:white">&nbsp;&nbsp;NMDS Plots</h2></div>

<br>

### PhyloGroup 2D NMDS

<br>
```{r med_nmds2d, echo=FALSE}
if ((params$asvMED == "true" || params$aminoMED == "true") && (params$type == "ASV" || params$type == "AminoType")){
    ##NMDS
    datax <- decostand(raredata,method="total") #method 'total' normalizes data to sum up to 1 --data5[,2:cols]

    MDS <- metaMDS(sqrt(datax),
                   distance = "bray",autotransform = FALSE,
                   k = 2,
                   maxit = 999,
                   trymax = params$try,
                   wascores = TRUE)

    if (MDS$converged == "TRUE") {

    data.scores <- as.data.frame(scores(MDS))
    data.scores$sample <- meddata5$sample
    data.scores.2 <- merge(data.scores, metadata, by="sample")

    p <- ggplot(data.scores.2, aes(x=NMDS1, y=NMDS2,color=treatment))+
      geom_point(size=2)+
      theme_classic()+
      theme(legend.title = element_blank())

    #fff <- ggplotly(p)
    #fff <- fff %>% layout(legend=list(y=.5))
    #fff

    fff <-plot_ly(data.scores.2, x=~NMDS1, y=~NMDS2, color=~treatment, colors=mycol, text = ~paste("Sample: ", sample))
    fff <- fff %>% layout(legend=list(y=.5))
    fff <- fff %>% config(toImageButtonOptions=list(format='svg',filename='2Dnmds', height= 500, width= 800, scale= 1))
    fff

    } else {
      print("NMDS did not converge. Printing PCoA")
      #calculate bray curtis distance
      bcdist <- vegdist(datax, "bray")
      res <- pcoa(bcdist)
      #res$values
      #biplot(res) #to make the boring 2D PCoA
      comp <- as.data.frame(res$vectors)
      comp$sample <- data5$sample
      comp2 <- merge(comp, metadata, by="sample")
      fig <- plot_ly(comp2, x = ~Axis.1, y = ~Axis.2, color= ~treatment, text = ~paste("Sample: ", sample,"<br>Treatment: ",treatment))
      fig <- fig %>% layout(xaxis = list(title = "PCoA 1"),
                            yaxis = list(title = "PCoA 2"))
      fig <- fig %>% layout(legend=list(y=.5))
      fig <- fig %>% config(toImageButtonOptions=list(format='svg',filename='2D_PostMED_PCoA', height= 500, width= 800, scale= 1))
      fig
    }
}
```

<br>
<br>
<br>
<br>

### PhyloGroup 3D NMDS

<br>

```{r med_nmds3d, echo=FALSE}
if ((params$asvMED == "true" || params$aminoMED == "true") && (params$type == "ASV" || params$type == "AminoType")){
    MDS3 <- metaMDS(sqrt(datax),
                   distance = "bray",autotransform = FALSE,
                   k = 3,
                   maxit = 999,
                   trymax = params$try,
                   wascores = TRUE)

    if (MDS3$converged == "TRUE") {

    data.scores3 <- as.data.frame(scores(MDS3))
    data.scores3$sample <- meddata5$sample
    data.scores.3 <- merge(data.scores3, metadata, by="sample")
    p3d <- plot_ly(data.scores.3, x=~NMDS1, y=~NMDS2, z=~NMDS3, text=~paste("Sample: ", sample),
            color=~treatment, colors=mycol,
            mode = 'markers', symbol = ~treatment, symbols = c('square','circle'),
            marker = list(opacity = .8,line=list(color = 'darkblue',width = 1))
            )
    p3d <- p3d %>% layout(legend=list(y=.5))
    p3d <- p3d %>% config(toImageButtonOptions=list(format='svg',filename='3Dnmds', height= 500, width= 800, scale= 1))
    p3d

    } else {
      print("NMDS did not converge. Printing PCoA")
      #calculate bray curtis distance
      bcdist <- vegdist(datax, "bray")
      res <- pcoa(bcdist)
      comp <- as.data.frame(res$vectors)
      comp$sample <- data5$sample
      comp2 <- merge(comp, metadata, by="sample")
      fig <- plot_ly(comp2, x = ~Axis.1, y = ~Axis.2, z = ~Axis.3, color= ~treatment, text = ~paste("Sample: ", sample,"<br>Treatment: ",treatment))
      fig <- fig %>% layout(xaxis = list(title = "PCoA 1"),
                            yaxis = list(title = "PCoA 2"),
                            zaxis = list(title = "PCoA 3"))
      fig <- fig %>% layout(legend=list(y=.5))
      fig <- fig %>% config(toImageButtonOptions=list(format='svg',filename='3D_PCoA', height= 500, width= 800, scale= 1))
      fig
    }
}
```

<br>
<br>

# Post Minimum Entropy Decomposition (MED) Analyses

<div class="rectangle"><h2 style="color:white">&nbsp;&nbsp;MED Group Breakdown Table</h2></div>

<br>
<br>

```{bash bash_med_table, include=FALSE}
if [[ $(ls | grep -wc "asv_medfile.csv") -eq 1 ]];then
  awk -F "," '{print $2}' asv_medfile.csv | sort | uniq | sort -g >group.list
  for x in $(cat group.list);do
    echo "${x}" >> ${x}.col
    grep -w "$x" asv_medfile.csv | awk -F "," '{print $1}' | sort >> ${x}.col
  done
  paste -d "," *.col > asv_medtable.csv
  rm *.col
else
    echo "Not MED"
fi

if [[ $(ls | grep -wc "amino_medfile.csv") -eq 1 ]];then
  awk -F "," '{print $2}' amino_medfile.csv | sort | awk -F "Group" '{print $2}' | sort -n | uniq > group.list
  for x in $(cat group.list);do
    gr="Group"${x}""
    echo "${gr}" > ${x}.col
    grep -w "$gr" amino_medfile.csv | awk -F "," '{print $1}' | sort >> ${x}.col
  done
  paste -d "," *.col > amino_medtable.csv
  rm *.col
else
      echo "Not MED"
fi
```

```{r med_table, echo=FALSE}
if (params$asvMED == "true" && params$type == "ASV"){
    #med_group=read.csv("asv_medfile.csv", header = F)
    #colnames(med_group) <- c("SequenceID", "Group", "Sequence")
    # change name of sequence to med peak or something
    #paged_table(med_group,options = list(rows.print = 20))
    med_group=read.csv("asv_medtable.csv", header = T)
    knitr::kable(med_group, digits = 2, align = 'c', booktabs = TRUE, caption = "Table: ASV MED Table") %>%
    kable_styling(font_size = 12, full_width = F)%>%
    scroll_box(width = "100%", height = "100%")
}

if (params$aminoMED == "true" && params$type == "AminoType"){
  #med_group=read.csv("amino_medfile.csv", header = F)
  #colnames(med_group) <- c("SequenceID", "Group", "Sequence")
  #paged_table(med_group,options = list(rows.print = 20))
  med_group=read.csv("amino_medtable.csv", header = T)
  knitr::kable(med_group, digits = 2, align = 'c', booktabs = TRUE, caption = "Table: Aminotype MED Table") %>%
  kable_styling(font_size = 12, full_width = F) %>%
  scroll_box(width = "100%", height = "100%")
}
```

<br>
<br>
<br>
<br>
<br>

<div class="rectangle"><h2 style="color:white">&nbsp;&nbsp;Post-MED Relative Abundance Plot</h2></div>

<br>
<br>

```{r med_plot, echo=FALSE}
if (params$asvMED == "true" && params$type == "ASV"){
  #medgroup=read.csv("uVP_AminoType_Groupingcounts.csv", check.names=FALSE)
  medgroup=read.csv("asv_groupcounts.csv", check.names=FALSE)
  meddata <- medgroup %>% select(-2)
  meddata2 <- aggregate(.~GroupID, meddata, sum)
  meddata3 <-as.data.frame(t(meddata2))
  meddata3$sample <- row.names(meddata3)
  colnames(meddata3)<- as.matrix(meddata3[1,])
  as.data.frame(meddata3)
  meddata3 <- meddata3[-1,]
  meddata3 <- meddata3 %>%
    rename(sample=GroupID)
  meddata3dim <- dim(meddata3)
  ##Loading metadata
  samples <- read.csv(sample_metadata, header = TRUE)
  #samples <- read.csv("rna_virus_meta.csv", header = TRUE)
  ##Combining data and metadata
  meddata4 <- merge(meddata3, samples, by="sample")
  #if "meddata4" removed it does not work
  #meddata4
  dim_meddata4 <- dim(meddata4)
  dim_samples <- dim(samples)
  cols <- dim_meddata4[2]-dim_samples[2]+1
  first <-colnames(meddata4)[2]
  last <- colnames(meddata4)[cols]
  meddata4[,2:cols] <- lapply(meddata4[,2:cols], as.character)
  meddata4[,2:cols] <- lapply(meddata4[,2:cols], as.numeric)
  #Calculate total reads per sample
  meddata5 <- meddata4%>%
    mutate(sum=select(.,2:cols)%>%
             apply(1, sum, na.rm=TRUE))
  #meddata5
  ##Filter samples with low reads
  nfil=params$minimumCounts
  #nfil=1000
  meddata5 <- meddata5 %>%
    filter(sum>nfil)
     #can cause errors
  meddata5dim <-dim(meddata5)
  minreads<-min(meddata5$sum)

  dataz <- decostand(meddata5[,2:cols],method="total") #method 'total' normalizes data to sum up to 1
  dataz$sample <- meddata5$sample
  metadata <- meddata5[,(cols+1):meddata5dim[2]]
  metadata$sample <- meddata5$sample
  datay <- merge(dataz, metadata, by="sample")
  datalong <- datay %>%
    tidyr::gather(first:last, key=hit, value=reads)
  ddd <- plot_ly(datalong, x=~sample, y=~reads, color=~hit, colors=mycol, type='bar')
  ddd <- ddd %>% layout(barmode = 'stack')
  ddd <- ddd %>% layout(legend = list(x=10,y=.5), xaxis=list(title = "Sample"), yaxis=list(title = "Relative abundance"))
  ddd <- ddd %>% config(toImageButtonOptions=list(format='svg',filename='Relative_abundance', height= 500, width= 800, scale= 1))
  ddd
}
if (params$aminoMED == "true" && params$type == "AminoType"){
  #medgroup=read.csv("uVP_AminoType_Groupingcounts.csv")
  medgroup=read.csv("amino_groupcounts.csv",  check.names=FALSE)
  meddata <- medgroup %>% select(-2)
  meddata2 <- aggregate(.~GroupID, meddata, sum)
  meddata3 <-as.data.frame(t(meddata2))
  meddata3$sample <- row.names(meddata3)
  colnames(meddata3)<- as.matrix(meddata3[1,])
  as.data.frame(meddata3)
  meddata3 <- meddata3[-1,]
  meddata3 <- meddata3 %>%
    rename(sample=GroupID)
  meddata3dim <- dim(meddata3)
  ##Loading metadata
  samples <- read.csv(sample_metadata, header = TRUE)
  ##Combining data and metadata
  meddata4 <- merge(meddata3, samples, by="sample")
  #if "meddata4" removed it does not work
  #meddata4
  dim_meddata4 <- dim(meddata4)
  dim_samples <- dim(samples)
  cols <- dim_meddata4[2]-dim_samples[2]+1
  first <-colnames(meddata4)[2]
  last <- colnames(meddata4)[cols]
  meddata4[,2:cols] <- lapply(meddata4[,2:cols], as.character)
  meddata4[,2:cols] <- lapply(meddata4[,2:cols], as.numeric)
  #Calculate total reads per sample
  meddata5 <- meddata4%>%
    mutate(sum=select(.,2:cols)%>%
             apply(1, sum, na.rm=TRUE))
  #meddata5
  ##Filter samples with low reads
  nfil=params$minimumCounts
  #nfil=1000
  meddata5 <- meddata5 %>%
    filter(sum>nfil)
     #can cause errors
  meddata5dim <-dim(meddata5)
  minreads<-min(meddata5$sum)

  dataz <- decostand(meddata5[,2:cols],method="total") #method 'total' normalizes data to sum up to 1
  dataz$sample <- meddata5$sample
  metadata <- meddata5[,(cols+1):meddata5dim[2]]
  metadata$sample <- meddata5$sample
  datay <- merge(dataz, metadata, by="sample")
  datalong <- datay %>%
    tidyr::gather(first:last, key=hit, value=reads)
  ddd <- plot_ly(datalong, x=~sample, y=~reads, color=~hit, colors=mycol, type='bar')
  ddd <- ddd %>% layout(barmode = 'stack')
  ddd <- ddd %>% layout(legend = list(x=10,y=.5), xaxis=list(title = "Sample"), yaxis=list(title = "Relative abundance"))
  ddd <- ddd %>% config(toImageButtonOptions=list(format='svg',filename='Relative_abundance', height= 500, width= 800, scale= 1))
  ddd
}
```

<br>
<br>
<br>
<br>
<br>

<div class="rectangle"><h2 style="color:white">&nbsp;&nbsp;Post-Med Rarefaction Curves</h2></div>

<br>
<br>

```{r med_rarefaction, echo=FALSE, cache=FALSE}
if ((params$asvMED == "true" || params$aminoMED == "true") && (params$type == "ASV" || params$type == "AminoType")){
##Rarefaction curves
rarefaction <- rarecurve(meddata5[,2:cols])

##rarefied dataset
raredata <- as.data.frame(rrarefy(meddata5[,2:cols], sample=minreads))
}
```

<br>
<br>
<br>
<br>
<br>
<br>

<div class="rectangle"><h2 style="color:white">&nbsp;&nbsp;Post-MED Diversity Analyses Plots</h2></div>

<br>
<br>

### Post-MED Shannon diversity

<br>
```{r med_diversity_analysis1a, echo=FALSE}
if ((params$asvMED == "true" || params$aminoMED == "true") && (params$type == "ASV" || params$type == "AminoType")){
    metadata <- meddata5[,(cols+1):meddata5dim[2]]
    metadata$sample <- meddata5$sample
    index <-diversity(raredata, index= "shannon")
    shannondata5 <- as.data.frame(index)
    shannondata5$sample<- meddata5$sample
    shannondata5_2 <- merge(shannondata5, metadata, by="sample")

    sh <- plot_ly(shannondata5_2, x=~treatment, y=~index, color=~treatment, colors=mycol, type="box", boxpoints = "all", pointpos = 0, jitter = 0.5)
    #sh <- sh %>% layout(title = list(text="Shannon diversty",y=.99))
    sh <- sh %>% layout(legend = list(x=10,y=.5), yaxis=list(title = "Index"), xaxis=list(title = "Treatment"))
    sh <- sh %>% config(toImageButtonOptions=list(format='svg',filename='ShannonDiv', height= 500, width= 800, scale= 1))
    sh
}
```

```{r med_diversity_analysis1b, echo=FALSE}
#divided because error
if ((params$asvMED == "true" || params$aminoMED == "true") && (params$type == "ASV" || params$type == "AminoType")){
    if (params$stats == "true" ) {
        shannonaov <- aov(index ~ treatment, data= shannondata5_2)
        st <- shapiro.test(resid(shannonaov))
        bt <- bartlett.test(index ~ treatment, data= shannondata5_2)

        if (st$p.value > .05 && bt$p.value > .05) {
          print("Shapiro Test of normality - data is normal p-value > 0.05")
          print(shapiro.test(resid(shannonaov)))
          writeLines("\n--------------------------------------------------------------\n")
          print("Bartlett Test variance homogeneity - variance is homogeneous p-value > 0.05")
          print(bartlett.test(index ~ treatment, data= shannondata5_2))
          writeLines("\n--------------------------------------------------------------\n")
          print("ANOVA Results")
          print(summary(shannonaov))
          writeLines("\n--------------------------------------------------------------\n")
          #Tukey Honest Significant Differences (pairwise comparison) - significant p <.05
          print("Tukey HSD - Pairwise comparison - significant differences indicated by p-value < 0.05")
          print(TukeyHSD(shannonaov))
          writeLines("\n--------------------------------------------------------------\n")
        } else {
          print("Shapiro Test of normality - data is normal if p-value > 0.05")
          print(shapiro.test(resid(shannonaov)))
          writeLines("\n--------------------------------------------------------------\n")
          print("Bartlett Test variance homogeneity - variance is homogeneous if p-value > 0.05")
          print(bartlett.test(index ~ treatment, data= shannondata5_2))
          writeLines("\n--------------------------------------------------------------\n")
          print("Data either not normal or variance not homogenous")
          print("Kruskal-Wallis Test - test significant if p <.05")
          #Kruskal-Wallis test - significant p <.05
          mykt <- kruskal.test(index ~ treatment, data= shannondata5_2)
          print(mykt)
          writeLines("\n--------------------------------------------------------------\n")
          if (mykt$p.value < .05) {
            #Pairwise comparison
            print("Wilcox.test - pairwise comparison")
            print(pairwise.wilcox.test(shannondata5_2$index, shannondata5_2$treatment, p.adjust.method = "BH"))
          } else {
            print("Data not significant. Skipping pairwise comparison")
          }
          writeLines("\n--------------------------------------------------------------\n")
          print("ANOVA - one or more of the assumptions not met, take with a grain of salt.")
          print(summary(shannonaov))
          writeLines("\n--------------------------------------------------------------\n")
        }
    } else {
        print("Stats skipped. To toggle on use \"--stats run\" in vAMPirus launch command")
    }
}
```
<br>
<br>
<br>
<br>

### Post-MED Simpson diversity

<br>
```{r med_diversity_analysis2a, echo=FALSE}
if ((params$asvMED == "true" || params$aminoMED == "true") && (params$type == "ASV" || params$type == "AminoType")){
    index <- diversity(raredata, index= "simpson")
    simpsondata5 <- as.data.frame(index)
    simpsondata5$sample<- meddata5$sample
    simpsondata5_2 <- merge(simpsondata5, metadata, by="sample")

    s <- plot_ly(simpsondata5_2, x=~treatment, y=~index, color=~treatment, colors=mycol, type="box", boxpoints = "all", pointpos = 0, jitter = 0.5)
    #s <- s %>% layout(title = list(text="Simpson diversty",y=.99))
    s <- s %>% layout(legend = list(x=10,y=.5), yaxis=list(title = "Index"), xaxis=list(title = "Treatment"))
    s <- s %>% config(toImageButtonOptions=list(format='svg',filename='SimpsonDiv', height= 500, width= 800, scale= 1))
    s
}
```

```{r med_diversity_analysis2b, echo=FALSE}
if ((params$asvMED == "true" || params$aminoMED == "true") && (params$type == "ASV" || params$type == "AminoType")){
    if (params$stats == "true" ) {
        simpsonaov <- aov(index ~ treatment, data= simpsondata5_2)
        st <- shapiro.test(resid(simpsonaov))
        bt <- bartlett.test(index ~ treatment, data= simpsondata5_2)

        if (st$p.value > .05 && bt$p.value > .05) {
          print("Shapiro Test of normality - data is normal p-value > 0.05")
          print(shapiro.test(resid(simpsonaov)))
          writeLines("\n--------------------------------------------------------------\n")
          print("Bartlett Test variance homogeneity - variance is homogeneous p-value > 0.05")
          print(bartlett.test(index ~ treatment, data= simpsondata5_2))
          writeLines("\n--------------------------------------------------------------\n")
          print("ANOVA Results")
          print(summary(simpsonaov))
          writeLines("\n--------------------------------------------------------------\n")
          #Tukey Honest Significant Differences (pairwise comparison) - significant p <.05
          print("Tukey HSD - Pairwise comparison - significant differences indicated by p-value < 0.05")
          print(TukeyHSD(simpsonaov))
          writeLines("\n--------------------------------------------------------------\n")
        } else {
          print("Shapiro Test of normality - data is normal if p-value > 0.05")
          print(shapiro.test(resid(simpsonaov)))
          writeLines("\n--------------------------------------------------------------\n")
          print("Bartlett Test variance homogeneity - variance is homogeneous if p-value > 0.05")
          print(bartlett.test(index ~ treatment, data= simpsondata5_2))
          writeLines("\n--------------------------------------------------------------\n")
          print("Data either not normal or variance not homogenous")
          print("Kruskal-Wallis Test - test significant if p <.05")
          #Kruskal-Wallis test - significant p <.05
          mykt <- kruskal.test(index ~ treatment, data= simpsondata5_2)
          print(mykt)
          writeLines("\n--------------------------------------------------------------\n")
          if (mykt$p.value < .05) {
            #Pairwise comparison
            print("Wilcox.test - pairwise comparison")
            print(pairwise.wilcox.test(simpsondata5_2$index, simpsondata5_2$treatment, p.adjust.method = "BH"))
          } else {
            print("Data not significant. Skipping pairwise comparison")
          }
          writeLines("\n--------------------------------------------------------------\n")
          print("ANOVA - one or more of the assumptions not met, take with a grain of salt.")
          print(summary(simpsonaov))
          writeLines("\n--------------------------------------------------------------\n")
        }
    } else {
        print("Stats skipped. To toggle on use \"--stats run\" in vAMPirus launch command")
    }
}
```
<br>
<br>
<br>
<br>

### Post-MED Richness

<br>
```{r med_diversity_analysis3a, echo=FALSE}
if ((params$asvMED == "true" || params$aminoMED == "true") && (params$type == "ASV" || params$type == "AminoType")){
    mind5<-min(meddata5$sum)
    index <- rarefy(meddata5[,2:cols], sample=mind5)
    rarerichnessdata5 <- as.data.frame(index)
    rarerichnessdata5$sample <-meddata5$sample
    richdata5_2 <- merge(rarerichnessdata5, metadata, by="sample")

    ri <- plot_ly(richdata5_2, x=~treatment, y=~index, color=~treatment, colors=mycol, type="box", boxpoints = "all", pointpos = 0, jitter = 0.5)
    #ri <- ri %>% layout(title = list(text="ASV Richness",y=.99))
    ri <- ri %>% layout(legend = list(x=10,y=.5), yaxis=list(title = "Index"), xaxis=list(title = "Treatment"))
    ri <- ri %>% config(toImageButtonOptions=list(format='svg',filename='SpeciesRich', height= 500, width= 800, scale= 1))
    ri
}
```

```{r med_diversity_analysis3b, echo=FALSE}
if ((params$asvMED == "true" || params$aminoMED == "true") && (params$type == "ASV" || params$type == "AminoType")){
    if (params$stats == "true" ) {
        richaov <- aov(index ~ treatment, data= richdata5_2)
        st <- shapiro.test(resid(richaov))
        bt <- bartlett.test(index ~ treatment, data= richdata5_2)

        if (st$p.value > .05 && bt$p.value > .05) {
          print("Shapiro Test of normality - data is normal p-value > 0.05")
          print(shapiro.test(resid(richaov)))
          writeLines("\n--------------------------------------------------------------\n")
          print("Bartlett Test variance homogeneity - variance is homogeneous p-value > 0.05")
          print(bartlett.test(index ~ treatment, data= richdata5_2))
          writeLines("\n--------------------------------------------------------------\n")
          print("ANOVA Results")
          print(summary(richaov))
          #Tukey Honest Significant Differences (pairwise comparison) - significant p <.05
          writeLines("\n--------------------------------------------------------------\n")
          print("Tukey HSD - Pairwise comparison - significant differences indicated by p-value < 0.05")
          print(TukeyHSD(richaov))
          writeLines("\n--------------------------------------------------------------\n")
        } else {
          print("Shapiro Test of normality - data is normal if p-value > 0.05")
          print(shapiro.test(resid(richaov)))
          writeLines("\n--------------------------------------------------------------\n")
          print("Bartlett Test variance homogeneity - variance is homogeneous if p-value > 0.05")
          print(bartlett.test(index ~ treatment, data= richdata5_2))
          writeLines("\n--------------------------------------------------------------\n")
          print("Data either not normal or variance not homogenous")
          print("Kruskal-Wallis Test - test significant if p <.05")
          #Kruskal-Wallis test - significant p <.05
          mykt <- kruskal.test(index ~ treatment, data= richdata5_2)
          print(mykt)
          writeLines("\n--------------------------------------------------------------\n")
          if (mykt$p.value < .05) {
            #Pairwise comparison
            print("Wilcox.test - pairwise comparison")
            print(pairwise.wilcox.test(richdata5_2$index, richdata5_2$treatment, p.adjust.method = "BH"))
          } else {
            print("Data not significant. Skipping pairwise comparison")
          }
          writeLines("\n--------------------------------------------------------------\n")
          print("ANOVA - one or more of the assumptions not met, take with a grain of salt.")
          print(summary(richaov))
          writeLines("\n--------------------------------------------------------------\n")
        }
    } else {
        print("Stats skipped. To toggle on use \"--stats run\" in vAMPirus launch command")
    }
}
```
<br>
<br>
<br>
<br>
<br>

<div class="rectangle"><h2 style="color:white">&nbsp;&nbsp;Post-MED Distance To Centroid</h2></div>
<br>
```{r med_distance2a, echo=FALSE}
if ((params$asvMED == "true" || params$aminoMED == "true") && (params$type == "ASV" || params$type == "AminoType")){
    ##Distance
    intermediate <- raredata
    bray.distance <- vegdist(sqrt(intermediate), method="bray")

    ##Dispersion
    disper <- betadisper(bray.distance, group = metadata$treatment, type="centroid")
    df <- data.frame(Distance_to_centroid=disper$distances,Group=disper$group)
    df$sample <- meddata5$sample
    df2 <- merge(df, metadata, by="sample")

    cen <- plot_ly(df2, x=~treatment, y=~Distance_to_centroid, color=~treatment, colors=mycol, type="box", boxpoints = "all", pointpos = 0, jitter = 0.5)
    #cen <- cen %>% layout(title = list(text="Distance to centroid",y=.99))
    cen <- cen %>% layout(legend = list(x=10,y=.5), yaxis=list(title = "Distance"), xaxis=list(title = "Treatment"))
    cen <- cen %>% config(toImageButtonOptions=list(format='svg',filename='Dispersion', height= 500, width= 800, scale= 1))
    cen
}
```
```{r med_distance2b, echo=FALSE}
if ((params$asvMED == "true" || params$aminoMED == "true") && (params$type == "ASV" || params$type == "AminoType")){
    if (params$stats == "true" ) {
        adn <- adonis(bray.distance~meddata5$treatment)
        adn
    } else {
        print("Stats skipped. To toggle on use \"--stats run\" in vAMPirus launch command")
    }
}
```
<br>
<br>
<br>
<br>
<br>

<div class="rectangle"><h2 style="color:white">&nbsp;&nbsp;NMDS Plots</h2></div>

<br>

### Post-MED 2D NMDS

<br>
```{r med_nmds2d, echo=FALSE}
if ((params$asvMED == "true" || params$aminoMED == "true") && (params$type == "ASV" || params$type == "AminoType")){
    ##NMDS
    datax <- decostand(raredata,method="total") #method 'total' normalizes data to sum up to 1 --data5[,2:cols]

    MDS <- metaMDS(sqrt(datax),
                   distance = "bray",autotransform = FALSE,
                   k = 2,
                   maxit = 999,
                   trymax = params$try,
                   wascores = TRUE)

    if (MDS$converged == "TRUE") {

    data.scores <- as.data.frame(scores(MDS))
    data.scores$sample <- meddata5$sample
    data.scores.2 <- merge(data.scores, metadata, by="sample")

    p <- ggplot(data.scores.2, aes(x=NMDS1, y=NMDS2,color=treatment))+
      geom_point(size=2)+
      theme_classic()+
      theme(legend.title = element_blank())

    #fff <- ggplotly(p)
    #fff <- fff %>% layout(legend=list(y=.5))
    #fff

    fff <-plot_ly(data.scores.2, x=~NMDS1, y=~NMDS2, color=~treatment, colors=mycol, text = ~paste("Sample: ", sample))
    fff <- fff %>% layout(legend=list(y=.5))
    fff <- fff %>% config(toImageButtonOptions=list(format='svg',filename='2Dnmds', height= 500, width= 800, scale= 1))
    fff

    } else {
      print("NMDS did not converge. Printing PCoA")
      #calculate bray curtis distance
      bcdist <- vegdist(datax, "bray")
      res <- pcoa(bcdist)
      #res$values
      #biplot(res) #to make the boring 2D PCoA
      comp <- as.data.frame(res$vectors)
      comp$sample <- data5$sample
      comp2 <- merge(comp, metadata, by="sample")
      fig <- plot_ly(comp2, x = ~Axis.1, y = ~Axis.2, color= ~treatment, text = ~paste("Sample: ", sample,"<br>Treatment: ",treatment))
      fig <- fig %>% layout(xaxis = list(title = "PCoA 1"),
                            yaxis = list(title = "PCoA 2"))
      fig <- fig %>% layout(legend=list(y=.5))
      fig <- fig %>% config(toImageButtonOptions=list(format='svg',filename='2D_PostMED_PCoA', height= 500, width= 800, scale= 1))
      fig
    }
}
```

<br>
<br>
<br>
<br>

### Post-MED 3D NMDS

<br>

```{r med_nmds3d, echo=FALSE}
if ((params$asvMED == "true" || params$aminoMED == "true") && (params$type == "ASV" || params$type == "AminoType")){
    MDS3 <- metaMDS(sqrt(datax),
                   distance = "bray",autotransform = FALSE,
                   k = 3,
                   maxit = 999,
                   trymax = params$try,
                   wascores = TRUE)

    if (MDS3$converged == "TRUE") {

    data.scores3 <- as.data.frame(scores(MDS3))
    data.scores3$sample <- meddata5$sample
    data.scores.3 <- merge(data.scores3, metadata, by="sample")
    p3d <- plot_ly(data.scores.3, x=~NMDS1, y=~NMDS2, z=~NMDS3, text=~paste("Sample: ", sample),
            color=~treatment, colors=mycol,
            mode = 'markers', symbol = ~treatment, symbols = c('square','circle'),
            marker = list(opacity = .8,line=list(color = 'darkblue',width = 1))
            )
    p3d <- p3d %>% layout(legend=list(y=.5))
    p3d <- p3d %>% config(toImageButtonOptions=list(format='svg',filename='3Dnmds', height= 500, width= 800, scale= 1))
    p3d

    } else {
      print("NMDS did not converge. Printing PCoA")
      #calculate bray curtis distance
      bcdist <- vegdist(datax, "bray")
      res <- pcoa(bcdist)
      comp <- as.data.frame(res$vectors)
      comp$sample <- data5$sample
      comp2 <- merge(comp, metadata, by="sample")
      fig <- plot_ly(comp2, x = ~Axis.1, y = ~Axis.2, z = ~Axis.3, color= ~treatment, text = ~paste("Sample: ", sample,"<br>Treatment: ",treatment))
      fig <- fig %>% layout(xaxis = list(title = "PCoA 1"),
                            yaxis = list(title = "PCoA 2"),
                            zaxis = list(title = "PCoA 3"))
      fig <- fig %>% layout(legend=list(y=.5))
      fig <- fig %>% config(toImageButtonOptions=list(format='svg',filename='3D_PCoA', height= 500, width= 800, scale= 1))
      fig
    }
}
```

<br>
<br>
<br>
<br>

<div class="rectangle"><h2 style="color:white">&nbsp;&nbsp;MED Group Representatives Phylogenetic Tree</h2></div>

<br>
<br>

```{r med_tree, echo=FALSE}
if (params$skipPhylogeny == "true") {
    # file fore here
    writeLines("\n--------------------------------------------------------------\n")
    cat(readLines(list.files(pattern="tree.txt")) , sep = '\n')
    writeLines("\n--------------------------------------------------------------\n")
} else if ((params$asvMED == "true" || params$aminoMED == "true") && (params$type == "ASV" || params$type == "AminoType")) {
    #tree=read.newick("vAMPrun_otu.55.raxml.support")
    tree=read.tree("grouptree.txt")
    p1 <- ggtree(tree)
    id <- tree$tip.label
    dat <- tibble::tibble(id = id)
    #cdat <- tibble::tibble(dcolor = c(1:247))
    metat <- p1$data %>% dplyr::inner_join(dat, c('label' = 'id'))

    p2 <- p1 + geom_point(data = metat, aes(x = x, y = y, label = id, color=label))
    ggplotly(p2, tooltip = "label")

    treelist=read.csv("tree_group.csv")
    colnames(treelist) <- c("GroupID","ID")

    metat2 <- metat %>% dplyr::inner_join(treelist, c('label' = 'GroupID'))
    p2 <- p1 + geom_point(data = metat2, aes(x = x, y = y, label = ID, color = label))
    ggplotly(p2, tooltip = "label")

}
```
This tree is a maximum likelihood tree made with the representative sequences from the MED groups formed in the oligotyping analysis using IQTREE2 and the parameters you specified in the vampirus.config file. Also, this is an interactive tree, you can zoom in and hover on nodes to know the sequence ID. For a better visualization of this tree, you can find the *.treefile with bootstrap support values within the results directory and visualize using programs like FigTree or ITOL.
<br>
<br>
<br>
<br>
<br>
<br>
<br>
<br>
<br>
